---
title: "04_WGCNA_SigMods"
author: "Samuel Gurr"
date: "2024-02-21"
output: pdf_document
---

### Set working directory
```{r setup, include=TRUE}
knitr::opts_chunk$set(echo = TRUE, 
                      warning = FALSE, 
                      message = FALSE, 
                      cache = TRUE)

knitr::opts_knit$set(root.dir = "C:/Users/samjg/Documents/Github_repositories/Airradians_CellularMolecular_OA/RAnalysis/Output/Transcriptomics/WGCNA/")
#knitr::opts_knit$set(root.dir = "C:/Users/samuel.gurr/Documents/Github_repositories/Airradians_CellularMolecular_OA/RAnalysis")
```

### Load libraries

```{r load_libraries, include = TRUE}
library(dplyr)
library(reshape2)
library(ggplot2)
library(tidyverse)
library(ggpubr)
```

## Reference data, KEGG Ids

**What is this?** 

We do not have KEGG annotation of the Airradians genome nor does cluster profiler recognize 
Airraidans entrezIDs to run. Therefore we used diamond to blastx the protein query Airradians genome to 
the Cvirginica annotation. Best blast hits determined by highest e value / bit score (same results) in order 
to connet Airradians gene ID to KEGG ids (of the oyster) to run cluster profiler

Below we have the Bgingica reference and the Airradians Cvirg diamon output. Load these to obtain connection of 
Airradians gene ID to KEGG ids to run cluster profiler

```{r diamond blastx and cvirg kegg reference}

Cvirg_KEGGIDs_Ref  <- read.csv(file = "../../../Data/Transcriptomics/metadata/seq_id_Cvirginica_master.csv",header = T) %>% 
                                dplyr::mutate(Cvirginica_KEGGID = paste0('cvn:',gsub(".*LOC", "", GeneID)),
                                              Cvirginica_TranscriptID = TranscriptID) %>% 
                                dplyr::rename(Cvirginica_GeneID    = GeneID,
                                              Cvirginica_ProteinID = ProteinID) %>% 
                                dplyr::select(Cvirginica_TranscriptID, Cvirginica_ProteinID, Cvirginica_GeneID, Cvirginica_KEGGID)

# diamond bastx output of Airradians genome with Cvirg query
Airr_Cvirg_diamond <- read.csv(file = "../../../Data/Transcriptomics/metadata/seq_id_AirrCvirg_MERGED_master.csv",header = T) %>% 
                                        dplyr::select(Airradians_TranscriptID, Airradians_GeneID, Airradians_ProteinID,
                                                      blastxEval_CvirgTranscriptID, blastxEval_CvirgProteinID, blastxEval_CvirgGeneID) %>% 
                                        dplyr::rename('Protein_name' = 'Airradians_ProteinID', 
                                                      'blastxEval_CvirgGeneID'    = 'blastxEval_CvirgProteinID',
                                                      'Protein_name_Cvirg'        = 'blastxEval_CvirgGeneID',
                                                      'Cvirginica_TranscriptID'   = 'blastxEval_CvirgTranscriptID',
                                                      'Cvirginica_GeneID'        = 'blastxEval_CvirgProteinID') %>% 
                                        dplyr::mutate(Cvirginica_KEGGID = paste0('cvn:',
                                                                           gsub('.*LOC','', Cvirginica_GeneID)))

```




## Load module membership files

* outputs from WGCNa for the cohort and challange read matrices

* truncate the files for only thos emodules that had significant correlations, review the heamap output from the 
WGCNa R script - in the output folder! 


```{r load module membership files}
load(file = "challenge/Challenge-networkConstruction-stepByStep.RData")
Challenge_ModuleMembership  <- read.csv(file="challenge/Challenge_WGCNA_ModulMembership.csv", 
                                        sep=',', header=TRUE) %>% 
                                # dplyr::select(Airradians.TranscriptID,Cvirginica.TranscriptID,moduleColor) %>% 
                                dplyr::filter(moduleColor %in% c('pink', 'magenta','red', 'brown')) %>% 
                                dplyr::rename(Airradians_TranscriptID = Airradians.TranscriptID,
                                              Cvirginica_TranscriptID = Cvirginica.TranscriptID) # to ease the merge to ref

load(file = "cohorts/Cohorts-networkConstruction-stepByStep.RData")
Cohort_ModuleMembership  <- read.csv(file="cohorts/Cohorts_WGCNA_ModulMembership.csv", 
                                        sep=',', header=TRUE) %>% 
                                # dplyr::select(Airradians.TranscriptID,Cvirginica.TranscriptID,moduleColor)
                                dplyr::filter(moduleColor %in% c('grey60', 'tan','midnightblue', 'purple', 'brown', 'black'))%>% 
                                dplyr::rename(Airradians_TranscriptID = Airradians.TranscriptID,
                                              Cvirginica_TranscriptID = Cvirginica.TranscriptID) # to ease the merge to ref


```

## Reduce module membership files by Stiduent P vlaue < 0.05

* WGNCA assigns transcripts to a pattern with 'membership' associated with pearsons corr coeff and students p value
below we truncate the module mebership (# genes per module) by sutdnts p value < 0.05 for that pattern

```{r Challenge module pvalue 0.05}

# Challenge pre cut off # genes per module 
Challenge_ModuleMembership %>% dplyr::group_by(moduleColor) %>%  dplyr::summarise(n=n())
# brown	1203			
# magenta	511			
# pink	530			
# red	619

Challenge_modCols   <- data.frame(moduleColor = unique(Challenge_ModuleMembership$moduleColor))
df_total            <- data.frame() # start dataframe 
Challenge_MM_0.05   <- data.frame(matrix(nrow = 1, ncol = 4)) # create dataframe to save cumunalitively during for loop
colnames(Challenge_MM_0.05) <- c('moduleColor', 'Airradians_TranscriptID','Cvirginica_TranscriptID','Protein_name') # names for comuns in the for loop


for (i in 1:nrow(Challenge_modCols)) {
  # start with loop by calling the row value common with the 'Master_KEGG_BPTerms' data frind from rbind above 
  modColor         <- Challenge_modCols[i,1]
  loopmodColor_cor <- paste("MM.", modColor, sep = '') # column name for mod color - PEarsons correlation value 
  loopmodColor_p   <- paste("p.MM.", modColor, sep = '') # column name for mod color - Students asymptotic p value 
  
  Mod_loop         <- Challenge_ModuleMembership %>% 
    dplyr::filter(moduleColor %in% modColor) %>% 
    dplyr::select(c('Airradians_TranscriptID',
                    'Cvirginica_TranscriptID',
                    'Protein_name',
                    'moduleColor', 
                    loopmodColor_cor,
                    loopmodColor_p))
  Mod_loop_0.05   <- as.data.frame(Mod_loop %>% 
                                     dplyr::filter(.[[6]] < 0.05) %>%
                           dplyr::select(!c(5,6)))
  
  
  # write csv file for the data reduced mod mem 
  
  loopdf             <- data.frame(Mod_loop_0.05) # name dataframe for this single row
  Challenge_MM_0.05  <- rbind(Challenge_MM_0.05,Mod_loop_0.05) #bind to a cumulative list dataframe
  print(Challenge_MM_0.05) # print to monitor progress
  
}


# Challenge post cut off # genes per module 
Challenge_MM_0.05 %>% tidyr::drop_na(moduleColor) %>%  dplyr::group_by(moduleColor) %>%  dplyr::summarise(n=n())
# brown	886			
# magenta	323			
# pink	431			
# red	359	
```

```{r Challenge plot by treatments}

# log transform the dds object
dds_1filt_rlogtrans <- as.data.frame(rlogTransformation(assay(dds_1filt))) # rlog transoform the expression data matrix (dds object)
dim(dds_1filt_rlogtrans) # 9866   30 - note there are 30 samples here, not ommitted from when run WGCNA
dds_1filt_rlog_trim <- dds_1filt_rlogtrans[,rownames(dds_1filt_master)] # overlap with the vst master to omit the 3 samples
dim(dds_1filt_rlog_trim) # 9866   27 - we now have the correct number of samples!
dds_1filt_rlog_master <- tibble::rownames_to_column(dds_1filt_rlog_trim,"Airradians.TranscriptID") # rownames as first column

# write out this rlog tranformed master data (used for plotting below1) 
write.csv(dds_1filt_rlog_master, file = "challenge/Challenge_rlog_transformed.csv")


# log transform the dds object
dds_1filt_vst <- as.data.frame(vst(assay(dds_1filt))) # rlog transoform the expression data matrix (dds object)
dim(dds_1filt_vst) # 9866   30 - note there are 30 samples here, not ommitted from when run WGCNA
dds_1filt_vst_trim <- dds_1filt_vst[,rownames(dds_1filt_master)] # overlap with the vst master to omit the 3 samples
dim(dds_1filt_vst_trim) # 9866   27 - we now have the correct number of samples!
dds_1filt_vst_master <- tibble::rownames_to_column(dds_1filt_vst_trim,"Airradians.TranscriptID") # rownames as first column

# write out this rlog tranformed master data (used for plotting below1) 
write.csv(dds_1filt_vst_master, file = "challenge/Challenge_vst_transformed.csv")


# call the module colors 
modcolor <- as.data.frame(unique(Challenge_MM_0.05$moduleColor) %>% na.omit())
names(modcolor)[1] <- "color"

library(ggpubr)
for(i in 1:nrow(modcolor)) {
  
  # vst read count date - narrow the columns - reshape and rename
  Mod_geneIDs     <- Challenge_MM_0.05 %>% 
                        dplyr::filter(moduleColor %in% modcolor[i,]) %>%  
                        dplyr::select("Airradians_TranscriptID") %>%  na.omit()
  vst_Mod        <- dds_1filt_vst_master %>% dplyr::filter(Airradians.TranscriptID %in% Mod_geneIDs[,1])
  vst_Mod_MELT   <- melt(vst_Mod, id=("Airradians.TranscriptID")) # melt using reshape2
  names(vst_Mod_MELT)[(2:3)] <-  c('Sample_num', 'vst_Expression') # change column names
  
  # merge by common row values 'Sample.Name'
  merged_Expdata_Mod <- merge(vst_Mod_MELT, exp_metadata_1_master, by ='Sample_num')
  
  # mean Exp response table 
  meanEXp_Mod <- merged_Expdata_Mod %>% 
    dplyr::select(c('Sample_num','vst_Expression','pCO2_history', 'pCO2_exposure', 'All_pCO2')) %>% 
    group_by(Sample_num, pCO2_history, pCO2_exposure, All_pCO2) %>%
    dplyr::summarize(mean.vstExp = mean(vst_Expression), 
                     sd.vstExp = sd(vst_Expression))
  
   
  # summarize datasets further by treatment period  =========================================================================================== #
  # remember:this is a mean of a mean!! First we complete mean vst exp by sample id (compiling all red module genes) - next all sample IDs by the treatment period (below
  # I will use these for mean SE plots 
  
  # pCO2 history ========================== #
  
  meanEXp_Summary.pCO2history <- meanEXp_Mod %>% 
                                  group_by(pCO2_history) %>%
                                  dplyr::summarize(mean = mean(mean.vstExp), 
                                                   sd = sd(sd.vstExp),
                                                   n = n(), 
                                                   se = sd/sqrt(n))
  
  # pCO2 exposure  ======================== #
  
  meanEXp_Summary.pCO2exposure <- meanEXp_Mod %>% 
                                group_by(pCO2_exposure) %>%
                                dplyr::summarize(mean = mean(mean.vstExp), 
                                                 sd = sd(sd.vstExp),
                                                 n = n(), 
                                                 se = sd/sqrt(n))
  
  # all pCO2 groups ======================= #
  
  meanEXp_Summary.All.pCO2 <- meanEXp_Mod %>% 
                                group_by(All_pCO2) %>%
                                dplyr::summarize(mean = mean(mean.vstExp), 
                                                 sd = sd(sd.vstExp),
                                                 n = n(), 
                                                 se = sd/sqrt(n))
  
 
  
  # PLOT =========================================================================================== #
  # The errorbars overlapped, so use position_dodge to move them horizontally
  pd <- position_dodge(0.3) # move them .05 to the left and right
   
  # Temperature mean sd plot ========================== #
  
  min_p1 <- min(meanEXp_Summary.pCO2history$mean) - max(meanEXp_Summary.pCO2history$se)
  max_p1 <- max(meanEXp_Summary.pCO2history$mean) + max(meanEXp_Summary.pCO2history$se)
  
  pCO2history.vst.Mod <- meanEXp_Summary.pCO2history %>% 
    dplyr::mutate(pCO2_history  = forcats::fct_relevel(pCO2_history, 'low', 'severe')) %>%
      ggplot(aes(x=pCO2_history, y=mean, fill=pCO2_history)) + 
      theme_classic() +
      geom_errorbar(aes(ymin=mean-se, ymax=mean+se), colour="black", width=.1, position=pd) +
      geom_line(position=pd) +
      geom_point(position=pd, size = 4, shape=21) +            
      xlab("pCO2_history") +
      ylab("vst gene expression") +        
      scale_fill_manual(values=c("#009E73", "#CC79A7")) +                  
      scale_y_continuous(limits=c((min_p1), (max_p1)), breaks= seq( floor(min_p1),ceiling(max_p1),by=0.05)) +
      theme(text = element_text(size=10), legend.position="none")
  
  
  # Salinity mean sd plot ========================== #
  
  min_p2 <- min(meanEXp_Summary.pCO2exposure$mean) - max(meanEXp_Summary.pCO2exposure$se)
  max_p2 <- max(meanEXp_Summary.pCO2exposure$mean) + max(meanEXp_Summary.pCO2exposure$se)
  
  pCO2exposure.vst.Mod <- meanEXp_Summary.pCO2exposure %>% 
    dplyr::mutate(pCO2_exposure = forcats::fct_relevel(pCO2_exposure, 'low', 'moderate', 'severe')) %>%
      ggplot(aes(x=pCO2_exposure, y=mean, shape=pCO2_exposure)) +
      theme_classic() +
      geom_errorbar(aes(ymin=mean-se, ymax=mean+se), colour="black", width=.1, position=pd) +
      geom_line(position=pd) +
      geom_point(position=pd, size = 4) +            
      xlab("pCO2_exposure") +
      ylab(NULL) +  
      scale_shape_manual(values=c(19,15,17)) +
        scale_y_continuous(limits=c((min_p2), (max_p2)), breaks= seq( floor(min_p2),ceiling(max_p2),by=0.05)) +
      theme(text = element_text(size=10), legend.position="none")
  
  
  # pCO2 mean sd plot ========================== #
  
  min_p3 <- min(meanEXp_Summary.All.pCO2$mean) - max(meanEXp_Summary.All.pCO2$se)
  max_p3 <- max(meanEXp_Summary.All.pCO2$mean) + max(meanEXp_Summary.All.pCO2$se)
  
  AllpCO2.vst.Mod <- meanEXp_Summary.All.pCO2 %>% 
    dplyr::mutate(pCO2_history  = c('low','low','low','severe','severe','severe')) %>%
    dplyr::mutate(pCO2_exposure = c('low','moderate','severe','low','moderate','severe')) %>%
        ggplot(aes(x=pCO2_exposure, y=mean, fill=pCO2_history, shape = pCO2_exposure)) +
        theme_classic() +
        geom_errorbar(aes(ymin=mean-se, ymax=mean+se), colour="black", width=.1, position=pd) +
        geom_line(position=pd) +
        geom_point(position=pd, size = 4) +            
        xlab("pCO2_exposure") +
        ylab("vst gene expression") + 
        ggtitle(paste0('Module ', modcolor[i,], ', N = ', nrow(Mod_geneIDs),' genes')) +
        scale_shape_manual(values=c(21,22,24)) +
        scale_fill_manual(values=c("#009E73", "#CC79A7")) +
        scale_y_continuous(limits=c((min_p3), (max_p3)), breaks= seq( floor(min_p3),ceiling(max_p3),by=0.05)) +
        theme(text = element_text(size=10), legend.position="none")

  
  # Assemble these together =========================================================================================== #
  #library(ggpubr)
  single.factor.plot <-  ggarrange( ggarrange(pCO2history.vst.Mod, pCO2exposure.vst.Mod, ncol=2),
                                    AllpCO2.vst.Mod,      
                                    plotlist = NULL,
                                    ncol = 1,
                                    nrow = 2,
                                    labels = NULL)
                          
  
  # output   ======================================================================================================== #
  pdf(paste("challenge/ModuleExpression_Treatment/SignificantModules_p0.05/Exp_Module_",modcolor[i,],".pdf", sep = ''),
      width=4, height=4)
  print(single.factor.plot)
  dev.off()
  
}


```


```{r Cohort module pvalue 0.05}

# Cohort pre cut off # genes per module 
Cohort_ModuleMembership %>% dplyr::group_by(moduleColor) %>%  dplyr::summarise(n=n())
# black	803			
# brown	997			
# grey60	259			
# midnightblue	304			
# purple	597			
# tan	534	

Cohort_modCols   <- data.frame(moduleColor = unique(Cohort_ModuleMembership$moduleColor))
df_total            <- data.frame() # start dataframe 
Cohort_MM_0.05   <- data.frame(matrix(nrow = 1, ncol = 4)) # create dataframe to save cumunalitively during for loop
colnames(Cohort_MM_0.05) <- c('moduleColor', 'Airradians_TranscriptID','Cvirginica_TranscriptID','Protein_name') # names for comuns in the for loop


for (i in 1:nrow(Cohort_modCols)) {
  # start with loop by calling the row value common with the 'Master_KEGG_BPTerms' data frind from rbind above 
  modColor         <- Cohort_modCols[i,1]
  loopmodColor_cor <- paste("MM.", modColor, sep = '') # column name for mod color - PEarsons correlation value 
  loopmodColor_p   <- paste("p.MM.", modColor, sep = '') # column name for mod color - Students asymptotic p value 
  
  Mod_loop         <- Cohort_ModuleMembership %>% 
    dplyr::filter(moduleColor %in% modColor) %>% 
    dplyr::select(c('Airradians_TranscriptID',
                    'Cvirginica_TranscriptID',
                    'Protein_name',
                    'moduleColor', 
                    loopmodColor_cor,
                    loopmodColor_p))
  Mod_loop_0.05   <- as.data.frame(Mod_loop %>% 
                                     dplyr::filter(.[[6]] < 0.05) %>%
                           dplyr::select(!c(5,6)))
  
  
  # write csv file for the data reduced mod mem 
  
  loopdf             <- data.frame(Mod_loop_0.05) # name dataframe for this single row
  Cohort_MM_0.05  <- rbind(Cohort_MM_0.05,Mod_loop_0.05) #bind to a cumulative list dataframe
  print(Cohort_MM_0.05) # print to monitor progress
  
}


# Cohort post cut off # genes per module 
Cohort_MM_0.05 %>% tidyr::drop_na(moduleColor) %>%  dplyr::group_by(moduleColor) %>%  dplyr::summarise(n=n())
# black	431			
# brown	602			
# grey60	210			
# midnightblue	221			
# purple	230			
# tan	402	
((803-431)/803) * 100 # 46.32628
((997-602)/997) * 100 # 39.61886
((259-210)/259) * 100 # 18.91892
((304-221)/304) * 100 # 27.30263
((597-230)/597) * 100 # 61.47404
((534-402)/534) * 100 # 24.7191
```

```{r Cohort plot by treatments}

# log transform the dds object
dds_2filt_rlogtrans <- as.data.frame(rlogTransformation(assay(dds_2filt))) # rlog transoform the expression data matrix (dds object)
dim(dds_2filt_rlogtrans) # 9866   30 - note there are 30 samples here, not ommitted from when run WGCNA
dds_2filt_rlog_trim <- dds_2filt_rlogtrans[,rownames(dds_2filt_master)] # overlap with the vst master to omit the 3 samples
dim(dds_2filt_rlog_trim) # 12163    14 - we now have the correct number of samples!
dds_2filt_rlog_master <- tibble::rownames_to_column(dds_2filt_rlog_trim,"Airradians.TranscriptID") # rownames as first column

# write out this rlog tranformed master data (used for plotting below1) 
write.csv(dds_2filt_rlog_master, file = "cohorts/Cohorts_rlog_transformed.csv")


# log transform the dds object
dds_2filt_vst <- as.data.frame(vst(assay(dds_2filt))) # rlog transoform the expression data matrix (dds object)
dim(dds_2filt_vst) # 12163    15- note there are 30 samples here, not ommitted from when run WGCNA
dds_2filt_vst_trim <- dds_2filt_vst[,rownames(dds_2filt_master)] # overlap with the vst master to omit the 3 samples
dim(dds_2filt_vst_trim) # 12163    14 - we now have the correct number of samples!
dds_2filt_vst_master <- tibble::rownames_to_column(dds_2filt_vst_trim,"Airradians.TranscriptID") # rownames as first column

# write out this rlog tranformed master data (used for plotting below1) 
write.csv(dds_2filt_vst_master, file = "cohorts/Cohorts_vst_transformed.csv")


# call the module colors 
modcolor <- as.data.frame(unique(Cohort_MM_0.05$moduleColor) %>% na.omit())
names(modcolor)[1] <- "color"

library(ggpubr)
for(i in 1:nrow(modcolor)) {
  
  # vst read count date - narrow the columns - reshape and rename
  Mod_geneIDs     <- Cohort_MM_0.05 %>% 
                        dplyr::filter(moduleColor %in% modcolor[i,]) %>%  
                        dplyr::select("Airradians_TranscriptID") %>%  na.omit()
  vst_Mod        <- dds_2filt_vst_master %>% dplyr::filter(Airradians.TranscriptID %in% Mod_geneIDs[,1])
  vst_Mod_MELT   <- melt(vst_Mod, id=("Airradians.TranscriptID")) # melt using reshape2
  names(vst_Mod_MELT)[(2:3)] <-  c('Sample_num', 'vst_Expression') # change column names
  
  # merge by common row values 'Sample.Name'
  merged_Expdata_Mod <- merge(vst_Mod_MELT, exp_metadata_2_master, by ='Sample_num')
  
  # mean Exp response table 
  meanEXp_Mod <- merged_Expdata_Mod %>% 
    dplyr::select(c('Sample_num','vst_Expression','pCO2_history', 'pCO2_exposure', 'All_pCO2')) %>% 
    group_by(Sample_num, pCO2_history, pCO2_exposure, All_pCO2) %>%
    dplyr::summarize(mean.vstExp = mean(vst_Expression), 
                     sd.vstExp = sd(vst_Expression))
  
   
  # summarize datasets further by treatment period  =========================================================================================== #
  # remember:this is a mean of a mean!! First we complete mean vst exp by sample id (compiling all red module genes) - next all sample IDs by the treatment period (below
  # I will use these for mean SE plots 
  
  # pCO2 history ========================== #
  
  meanEXp_Summary.pCO2history <- meanEXp_Mod %>% 
                                  group_by(pCO2_history) %>%
                                  dplyr::summarize(mean = mean(mean.vstExp), 
                                                   sd = sd(sd.vstExp),
                                                   n = n(), 
                                                   se = sd/sqrt(n))
  
  # pCO2 exposure  ======================== #
  
  meanEXp_Summary.pCO2exposure <- meanEXp_Mod %>% 
                                group_by(pCO2_exposure) %>%
                                dplyr::summarize(mean = mean(mean.vstExp), 
                                                 sd = sd(sd.vstExp),
                                                 n = n(), 
                                                 se = sd/sqrt(n))
  
  # all pCO2 groups ======================= #
  
  meanEXp_Summary.All.pCO2 <- meanEXp_Mod %>% 
                                group_by(All_pCO2) %>%
                                dplyr::summarize(mean = mean(mean.vstExp), 
                                                 sd = sd(sd.vstExp),
                                                 n = n(), 
                                                 se = sd/sqrt(n))
  
 
  
  # PLOT =========================================================================================== #
  # The errorbars overlapped, so use position_dodge to move them horizontally
  pd <- position_dodge(0.3) # move them .05 to the left and right
   
  # Temperature mean sd plot ========================== #
  
  min_p1 <- min(meanEXp_Summary.pCO2history$mean) - max(meanEXp_Summary.pCO2history$se)
  max_p1 <- max(meanEXp_Summary.pCO2history$mean) + max(meanEXp_Summary.pCO2history$se)
  
  pCO2history.vst.Mod <- meanEXp_Summary.pCO2history %>% 
    dplyr::mutate(pCO2_history  = forcats::fct_relevel(pCO2_history, 'low', 'moderate', 'severe')) %>%
      ggplot(aes(x=pCO2_history, y=mean, fill=pCO2_history)) + 
      theme_classic() +
      geom_errorbar(aes(ymin=mean-se, ymax=mean+se), colour="black", width=.1, position=pd) +
      geom_line(position=pd) +
      geom_point(position=pd, size = 4, shape=21) +            
      xlab("pCO2_history") +
      ylab("vst gene expression") +        
      scale_fill_manual(values=c("#009E73", "#E69F00","#CC79A7")) +                  
      scale_y_continuous(limits=c((min_p1), (max_p1)), breaks= seq( floor(min_p1),ceiling(max_p1),by=0.05)) +
      theme(text = element_text(size=10), legend.position="none")
  
                          
  
  # output   ======================================================================================================== #
  pdf(paste("cohorts/ModuleExpression_Treatment/SignificantModules_p0.05/Exp_Module_",modcolor[i,],".pdf", sep = ''),
      width=4, height=4)
  print(pCO2history.vst.Mod)
  dev.off()
  
}


```



### Merge Cvirginica reference with reduced module membership files

* Airradians does not have recognizale GO and KEGG ids for g profiler (onling GO enrichment platform)
therefore we ran diamon (blastx) against the Cvirginica protein uwery to obtain functional annotation 
for *most* Airradians genes - we report the cover in the manuscript, some NAs others uncharacterized! 

* **what to do with these outputs?** Tese output C vriginica gene IDs (LOC...)

```{r merge GO and KEGG info}

# merge and write out 

Challenge_MM_annotated <- merge( (Challenge_MM_0.05 %>% tidyr::drop_na(moduleColor)),
                              Cvirg_KEGGIDs_Ref, 
                              by = 'Cvirginica_TranscriptID', all=T) %>% 
                              dplyr::arrange(moduleColor)
nrow(Challenge_MM_annotated %>% dplyr::filter(moduleColor %in% 'red'))
write.csv(Challenge_MM_annotated, "challenge/Challenge_SignificantModules_Annotated.csv")


Cohort_MM_annotated <- merge( (Cohort_MM_0.05 %>% tidyr::drop_na(moduleColor)),
                              Cvirg_KEGGIDs_Ref, 
                              by = 'Cvirginica_TranscriptID', all=T) %>% 
                              dplyr::arrange(moduleColor)
write.csv(Cohort_MM_annotated, "cohorts/Cohorts_SignificantModules_Annotated.csv")
```

