---
title: "DESeq2"
author: "Samuel Gurr"
date: "2024-02-21"
output: pdf_document
---


### Set working directory
```{r setup, include=TRUE}
knitr::opts_chunk$set(echo = TRUE, 
                      warning = FALSE, 
                      message = FALSE, 
                      cache = TRUE)

knitr::opts_knit$set(root.dir = "C:/Users/samjg/Documents/Github_repositories/Airradians_CellularMolecular_OA/RAnalysis")
knitr::opts_knit$set(root.dir = "C:/Users/samuel.gurr/Documents/Github_repositories/Airradians_CellularMolecular_OA/RAnalysis")
```

### Load libraries
```{r load_libraries, include = TRUE}
# load libraries - notes show the install command needed to install (pre installed)
library(tidyr)
library(DESeq2) # note: this was previously installed with the command `BiocManager::install("DESeq2")`
library(edgeR)
library(goseq)
library(dplyr)
library(GenomicFeatures)
library(data.table)
library(calibrate)
library(affycoretools) # note: this was previously installed with the BiocManager::install("affycoretools")
library(data.table)
library(vsn)
# Plotting
library(ggplot2)
library(cowplot)
library(pheatmap)
library(gplots)
library(RColorBrewer)
library(EnhancedVolcano)  # note: this was previously installed with the command `BiocManager::install("tweeDEseq")`
library(tweeDEseq)
library(pcaExplorer) # note: this was previously installed with the command `BiocManager::install("pcaExplorer")
library(ggpubr)
library(stringr)
```

### Sample metadata - Experimental treatments/groups
```{r experiment_data, include = TRUE}
### experiment metadata [from Count_Matrix_Stats.Filter.R]  - convert characaters to factors for DESeq2
exp_metadata <- read.csv(file="Data/Transcriptomics/metadata/metadata.csv", sep=',', header=TRUE) %>% 
                    dplyr::mutate(All_pCO2 = paste0(pCO2_history, pCO2_exposure))

# NOTE: the Smaple_num S32 should be omitted - it entials a single individual from severe pCO2 history 
# exposure to moderate pCO2 - omit from both the expreiment metadata and from the count amtrices for building dds objects! 
exp_metadata <- exp_metadata %>% dplyr::filter(!Sample_num == 'S32')# 20230516	80	14	severe	moderate	C7	S32
nrow(exp_metadata) == 35 # TRUE!!
```

### Gene count data 

* below are the raw and filtered read matrices

* 35 samples as the following 

  - Low pCO2 history (N =15), under low, moderate, and high pCO2 exposure (N = 5 ea)
  
  - Moderate pCO2 history (N = 5), only under *matched* moderate pCO2 exposure
    
  - High pCO2 history (N =15), under low, moderate, and high pCO2 exposure (N = 5 ea)

* two questions to take with WGCNA 

  - (1) Response to the challenge, pin low v high history under all challenges (N = 30) 
  
  - (2) Cohort differences under their *matched* exposure (N = 15)

### Read in the raw and filtered data

- 'Raw' - unfiltered read count matrix, formatted as a dataframe

  * raw_counts = raw unfiltered (N = 35 samples)
  
- 'filtered' - using edgeR and CPM thresholds to target mean 1000 reads per gene

  * filtered_counts_all = N = 35; 5 cpm used to acquire mean 1000 reads per gene in 50% samples
  
  * filtered_counts_challenge = N = 30; 5 cpm used to acquire mean 1000 reads per gene in 50% samples
  
  * filtered_counts_cohort = N = 15; 9 cpm used to acquire mean 1000 reads per gene in 33.33% samples
```{r call the count matrices, include = TRUE}
# non-filtered read matrix
raw_counts     <-
                  read.csv(file="Output/Transcriptomics/raw_count_matrix_all.csv", 
                                                               sep=',', 
                                                               header=TRUE)  %>% 
                      dplyr::rename(transcript_id = X) %>% 
                      tibble::column_to_rownames(var = "transcript_id") %>% 
                      dplyr::select(!S32) %>% 
                      as.matrix()  
ncol(raw_counts) # 35 - good! 
nrow(raw_counts) # 26686 - all genes
# gather the 3CPM filtered read matrices 

# all filtered
filtered_counts_all <-
                  read.csv(file="Output/Transcriptomics/Filtered_count_matrix/filtered_count_matrix_all.csv", 
                                                               sep=',', 
                                                               header=TRUE)  %>% 
                      dplyr::rename(transcript_id = X) %>% 
                      tibble::column_to_rownames(var = "transcript_id") %>% 
                      dplyr::select(!S32) %>% 
                      as.matrix()  
ncol(filtered_counts_all) # 35, good! 
nrow(filtered_counts_all) # 8216 subset from the 5CPM cutoff


# challenge
filtered_counts_challenge <-
                  read.csv(file="Output/Transcriptomics/Filtered_count_matrix/filtered_count_matrix_challenge.csv", 
                                                               sep=',', 
                                                               header=TRUE)  %>% 
                      dplyr::rename(transcript_id = X) %>% 
                      tibble::column_to_rownames(var = "transcript_id") %>% 
                      as.matrix()  
ncol(filtered_counts_challenge) # 30, good! 
nrow(filtered_counts_challenge) # 8113 subset from the 5CPM cutoff

# cohort
filtered_counts_cohort <-
                  read.csv(file="Output/Transcriptomics/Filtered_count_matrix/filtered_count_matrix_cohort.csv", 
                                                               sep=',', 
                                                               header=TRUE)  %>% 
                      dplyr::rename(transcript_id = X) %>% 
                      tibble::column_to_rownames(var = "transcript_id") %>% 
                      as.matrix()  
ncol(filtered_counts_cohort) # 15, good! 
nrow(filtered_counts_cohort) # 8654 subset from the 9CPM cutoff

```




```{r filtered cohort data boxplots visual}
filtered_counts_cohor_df <- (as.data.frame((filtered_counts_cohort)))
str(filtered_counts_cohor_df, strict.width = "wrap")
data_st                  <- as.data.frame(t(apply(as.matrix(filtered_counts_cohor_df), 1, scale)))
str(data_st, strict.width = "wrap")
data_st                  <- tibble::rownames_to_column(data_st, "ID")
colnames(data_st)[2:ncol(data_st)] <- colnames(filtered_counts_cohor_df)

nID <- which(is.nan(data_st[,2]))
data_st[nID,2:length(data_st)] <- 0

str(data_st, strict.width = "wrap")

library(reshape2)
ggplot(data = melt(data_st), aes(x=variable, y=value)) + geom_boxplot()

# looks like sample S.33 is poor here - we should ommit
# Note: downsteram without omitting we saw this is an outliter in PCA analysis too 
colnames(data_st)[2:6] <- c('Low.1','Low.2','Low.3','Low.4','Low.5')
colnames(data_st)[7:11] <- c('Mod.1','Mod.2','Mod.3','Mod.4','Mod.5')
colnames(data_st)[12:16] <- c('High.1','High.2','High.3','High.4','High.5')
```

### (1) Response to challenge - read count (raw and filtered) and metadata
* ncol() == 30!
* here we simply call the filtered data '1' for challenge
```{r edit data for #1}
exp_metadata_1    <- exp_metadata %>% dplyr::filter(pCO2_history %in% c('low','severe')) # omit moderate history
raw_counts_1      <- raw_counts[,exp_metadata_1$Sample_num] # omit columns from the raw matrix
filtered_counts_1 <- filtered_counts_challenge # use the filtered data - already catered to the target samples when filtered prior
```

### (2) Cohort differences under *match* - read count (raw and filtered) and metadata
* here we simply call the filtered data '2' for cohort
* omit sample num S33 due to the standardized being off and the downsteram PCA shows it is off
* ncol() == 14!
```{r edit data for #2}
exp_metadata_2    <- exp_metadata %>% dplyr::filter(All_pCO2 %in% c('lowlow', 'moderatemoderate', 'severesevere')) %>% 
                        dplyr::filter(!Sample_num == 'S36')
raw_counts_2      <- raw_counts[,exp_metadata_2$Sample_num]
filtered_counts_2 <- filtered_counts_cohort[,exp_metadata_2$Sample_num] # use the filtered data - already catered to the target samples when filtered prior
ncol(filtered_counts_2)
```


### DESeqDataSet or 'dds' object (using DESeq2) 
```{r build dds object}
# (1) Response to challenge 
dds_1raw <- DESeqDataSetFromMatrix(countData = raw_counts_1,
                                 colData = exp_metadata_1, design = ~pCO2_history+pCO2_exposure) # DESeq Data Set (dds)

dds_1filt <- DESeqDataSetFromMatrix(countData = filtered_counts_1,
                                 colData = exp_metadata_1, design = ~pCO2_history+pCO2_exposure) # DESeq Data Set (dds)

# (2) Cohort differences under *match*
dds_2raw <- DESeqDataSetFromMatrix(countData = raw_counts_2,
                                 colData = exp_metadata_2, design = ~pCO2_history) # DESeq Data Set (dds)

dds_2filt <- DESeqDataSetFromMatrix(countData = filtered_counts_2,
                                 colData = exp_metadata_2, design = ~pCO2_history) # DESeq Data Set (dds)
```


### run DESeq model - Full & Matched

```{r, Run DESeq}
# Full models
# (1) Response to challenge 

dds_1raw       <- DESeq(dds_1raw) # 
dds_1filt      <- DESeq(dds_1filt) # 

# (2) Cohort differences under *match*
dds_2raw     <- DESeq(dds_2raw) # 
dds_2filt    <- DESeq(dds_2filt) #
```

## Transform Exp data 

* Why?
 - the transformed data can be used to run simple PCA 
 - identify whether samples appear as outliers sufficient for omission and rerun dds.run

```{r expression data transformation}

# challenge 
vstExp.raw_challenge       <- vst(dds_1raw)
vstExp.raw_challenge.assay <- assay(vstExp.raw_challenge) # call only the transformed coutns in the dds object
vstExp.raw_challenge.df    <- as.data.frame(vstExp.raw_challenge.assay)
vstExp.raw_challenge.df    <- tibble::rownames_to_column(vstExp.raw_challenge.df, var = "Gene")


write.csv(vstExp.raw_challenge.df, "Output/Transcriptomics/vstExp_raw_count_matrix_challenge.csv")



vstExp.filt_challenge       <- vst(dds_1filt)
vstExp.filt_challenge.assay <- assay(vstExp.filt_challenge) # call only the transformed coutns in the dds object
vstExp.filt_challenge.df    <- as.data.frame(vstExp.filt_challenge.assay)
vstExp.filt_challenge.df    <- tibble::rownames_to_column(vstExp.filt_challenge.df, var = "Gene")


write.csv(vstExp.filt_challenge.df, "Output/Transcriptomics/vstExp_filtered_count_matrix_challenge.csv")



# cohort  
vstExp.raw_cohort       <- vst(dds_2raw)
vstExp.raw_cohort.assay <- assay(vstExp.raw_cohort) # call only the transformed coutns in the dds object
vstExp.raw_cohort.df    <- as.data.frame(vstExp.raw_cohort.assay)
vstExp.raw_cohort.df    <- tibble::rownames_to_column(vstExp.raw_cohort.df, var = "Gene")
vstExp.raw_cohort.df    <- vstExp.raw_cohort.df %>% 
                              mutate_at(c(2:ncol(vstExp.raw_cohort.df)), as.numeric)

write.csv(vstExp.raw_cohort.df, "Output/Transcriptomics/vstExp_raw_count_matrix_cohort.csv")



vstExp.filt_cohort       <- vst(dds_2filt)
vstExp.filt_cohort.assay <- assay(vstExp.filt_cohort) # call only the transformed coutns in the dds object
vstExp.filt_cohort.df    <- as.data.frame(vstExp.filt_cohort.assay)
vstExp.filt_cohort.df    <- tibble::rownames_to_column(vstExp.filt_cohort.df, var = "Gene")
vstExp.filt_cohort.df    <- vstExp.filt_cohort.df %>% 
                              mutate_at(c(2:ncol(vstExp.filt_cohort.df)), as.numeric)

write.csv(vstExp.filt_cohort.df, "Output/Transcriptomics/vstExp_filtered_count_matrix_cohort.csv")

```


## Visualize data using scatterplot matrices
```{r Scatterplot matrix}
# fxns
reg <- function(x, y, ...) {
  points(x,y, ...)
  abline(lm(y~x)) 
  }# made to draw regression line instead of lowess line
panel.cor <- function(x, y, digits = 2, prefix = "", cex.cor, ...)
{
  usr <- par("usr"); on.exit(par(usr))
  par(usr = c(0, 1, 0, 1))
  r <- abs(cor(x, y))
  txt <- format(c(r, 0.123456789), digits = digits)[1]
  txt <- paste0(prefix, txt)
  text(0.5, 0.5, txt, cex = 1.1, font = 4)
}

# now you can run PAirs with correct formatting 

vstExp.filt_cohort.PLOTTING <- vstExp.filt_cohort.df
colnames(vstExp.filt_cohort.PLOTTING)[2:6] <- c('Low.1','Low.2','Low.3','Low.4','Low.5')
colnames(vstExp.filt_cohort.PLOTTING)[7:11] <- c('Mod.1','Mod.2','Mod.3','Mod.4','Mod.5')
colnames(vstExp.filt_cohort.PLOTTING)[12:15] <- c('High.1','High.2','High.3','High.4')

## Low v Moderate  - filtered data pinning 3 from low and 3 from moderate 
pairs(vstExp.filt_cohort.PLOTTING[c(4,5,6,7,9,10)], 
      upper.panel = reg, # replace HERE for panel.smooth #
      cex = 1.5, 
      pch = 19, 
      col = adjustcolor(4, .4), 
      cex.labels = 2, 
      font.labels = 2, 
      lower.panel = panel.cor)
## Moderate v High - filtered data pinning 4 from low and 3 from high 
pairs(vstExp.filt_cohort.PLOTTING[c(7,9,10,12,13,14)], 
                          upper.panel = reg, # replace HERE for panel.smooth #
                          cex = 1.5, pch = 19, 
                          col = adjustcolor(4, .4), 
                          cex.labels = 2, 
                          font.labels = 2, 
                          lower.panel = panel.cor)
## Low v High - filtered data pinning 4 from low and 3 from high 
pairs(vstExp.filt_cohort.PLOTTING[c(4,5,6,12,13,14)], 
                          upper.panel = reg, # replace HERE for panel.smooth #
                          cex = 1.5, pch = 19, 
                          col = adjustcolor(4, .4), 
                          cex.labels = 2, 
                          font.labels = 2, 
                          lower.panel = panel.cor)

```


## Run diagnostic plots
```{r, Diagnostic plotting}
library(ggplot2)
vstExp.filt_cohort.df
hist(vstExp.filt_cohort.df)
meanSdPlot(vstExp.filt_cohort.df)

pcaData_vstExp_2    <- plotPCA(vstExp.filt_cohort, 
                               intgroup = "pCO2_history", 
                               returnData = TRUE)
percentVar_vstExp_2 <- round(100 * attr(pcaData_vstExp_2, "percentVar"))
ggplot(pcaData_vstExp_2, aes(x = PC1, 
                             y = PC2, 
                             color = pCO2_history, 
                             label=name)) +

      geom_point(size =3) +
      theme_classic() +
      stat_ellipse() +
      #theme(text = element_text(size=15)) +
      theme_classic() +
      theme(text = element_text(size=15)) +
      ggtitle("Cohort PCA: vst expression data") +
      xlab(paste0("PC1: ", percentVar_vstExp_2[1], "% variance")) + 
      ylab(paste0("PC2: ", percentVar_vstExp_2[2], "% variance")) +
      coord_fixed()

```


## DEGs - data tables
* Examine DEGs how many? what are they?

### Cohorts data 
```{r ohorts DEGsLow v. Severe}

# Edit the annotation file (Cvirg blasted to annotate draft genome contigs of Airradians)
Airr_Cvirg_annotation <- read.csv(file="Data/Transcriptomics/metadata/seq_id_AirrCvirg_MERGED_master.csv",
                                  sep = ',', 
                                  header = T) %>% 
                        dplyr::select(c('Airradians_TranscriptID',
                                 "blastxEval_CvirgTranscriptID",
                                 "blastxEval_CvirgProteinID",
                                 "blastxEval_CvirgGeneID",
                                 "blastxEval_CvirgGOterms"))

# ========================================================== 
#  CO2 Cohorts - Low v. High   
# ==========================================================
# Grab DEGs from  the 'dds.run' High
DEGresults_pCO2History.LowHigh         <- results(dds_2filt, 
                                                  contrast=c("pCO2_history", 
                                                             "low", "severe"),
                                                  alpha = 0.05) # FDR is the alpha 0.05

numDEGs.pval_pCO2History.LowHigh       <- data.frame(table(
                                                      DEGresults_pCO2History.LowHigh$pvalue<0.05)
                                                     )[2,2] # DEGs - NOT considering LFC - just p adj


numDEGs.padj_pCO2History.LowHigh       <- data.frame(table(DEGresults_pCO2History.LowHigh$padj<0.05))[2,2] 


DEGresults.ordered_pCO2History.LowHigh <- DEGresults_pCO2History.LowHigh[order(
                                            DEGresults_pCO2History.LowHigh$padj), ] # Order by adjusted p-value

num.UpReg_pCO2History.LowHigh          <- sum((DEGresults.ordered_pCO2History.LowHigh$log2FoldChange
                                               [1:numDEGs.padj_pCO2History.LowHigh] > 0) == TRUE) #  LFC >= 1

num.DownReg_pCO2History.LowHigh        <- sum((DEGresults.ordered_pCO2History.LowHigh$log2FoldChange
                                               [1:numDEGs.padj_pCO2History.LowHigh] < 0) == TRUE) # LFC >= 1

total_pCO2History.LowHigh              <- sum(num.UpReg_pCO2History.LowHigh,num.DownReg_pCO2History.LowHigh) # sum of DEGs with the criteria pdj < 0.05 + LFC>1 (< -1)
# Write results - covert to as.data.frame for the ordered results
resdata_pCO2History.LowHigh            <- merge(as.data.frame(DEGresults.ordered_pCO2History.LowHigh), 
                                                as.data.frame(counts(dds_2filt, normalized=TRUE)), 
                                                by="row.names", 
                                                sort=FALSE) %>% dplyr::filter(padj < 0.05)## Merge with normalized count data
names(resdata_pCO2History.LowHigh)[1] <- "Airradians_TranscriptID"
resdata.annot_pCO2History.LowHigh     <- merge(Airr_Cvirg_annotation,resdata_pCO2History.LowHigh, 
                                               by = 'Airradians_TranscriptID') 
# View(resdata.annot_pCO2History.LowHigh)
nrow(resdata.annot_pCO2History.LowHigh) == total_pCO2History.LowHigh
resdata.annot_pCO2History.LowHigh     <- resdata.annot_pCO2History.LowHigh[order(
                                             resdata.annot_pCO2History.LowHigh$padj),]  # Order by adjusted p-value
# write out annotated DESeq2 DEGs
write.csv(resdata.annot_pCO2History.LowHigh, "Output/Transcriptomics/DESeq2/pCO2History_Low_v_High.csv") # write


# ========================================================== 
#  CO2 Cohorts - Low v. Moderate   
# ==========================================================
# Grab DEGs from  the 'dds.run' Moderate
DEGresults_pCO2History.LowModerate         <- results(dds_2filt, 
                                                  contrast=c("pCO2_history", 
                                                             "low", "moderate"),
                                                  alpha = 0.05) # FDR is the alpha 0.05

numDEGs.pval_pCO2History.LowModerate       <- data.frame(table(
                                                      DEGresults_pCO2History.LowModerate$pvalue<0.05)
                                                     )[2,2] # DEGs - NOT considering LFC - just p adj


numDEGs.padj_pCO2History.LowModerate       <- data.frame(table(
                                               DEGresults_pCO2History.LowModerate$padj<0.05))[2,2] 


DEGresults.ordered_pCO2History.LowModerate <- DEGresults_pCO2History.LowModerate[order(
                                                  DEGresults_pCO2History.LowModerate$padj), ] # Order by adjusted p-value

num.UpReg_pCO2History.LowModerate          <- sum((DEGresults.ordered_pCO2History.LowModerate$log2FoldChange
                                               [1:numDEGs.padj_pCO2History.LowModerate] > 0) == TRUE) #  LFC >= 1

num.DownReg_pCO2History.LowModerate        <- sum((DEGresults.ordered_pCO2History.LowModerate$log2FoldChange
                                               [1:numDEGs.padj_pCO2History.LowModerate] < 0) == TRUE) # LFC >= 1

total_pCO2History.LowModerate              <- sum(num.UpReg_pCO2History.LowModerate,num.DownReg_pCO2History.LowModerate) # sum of DEGs with the criteria pdj < 0.05 + LFC>1 (< -1)
# Write results - covert to as.data.frame for the ordered results
resdata_pCO2History.LowModerate            <- merge(as.data.frame(DEGresults.ordered_pCO2History.LowModerate), 
                                                as.data.frame(counts(dds_2filt, normalized=TRUE)), 
                                                by="row.names", 
                                                sort=FALSE) %>% dplyr::filter(padj < 0.05)## Merge with normalized count data
names(resdata_pCO2History.LowModerate)[1] <- "Airradians_TranscriptID"
resdata.annot_pCO2History.LowModerate     <- merge(Airr_Cvirg_annotation,resdata_pCO2History.LowModerate, 
                                               by = 'Airradians_TranscriptID') 
# View(resdata.annot_pCO2History.LowModerate)
nrow(resdata.annot_pCO2History.LowModerate) == total_pCO2History.LowModerate
resdata.annot_pCO2History.LowModerate     <- resdata.annot_pCO2History.LowModerate[order(
                                             resdata.annot_pCO2History.LowModerate$padj),]  # Order by adjusted p-value
# write out annotated DESeq2 DEGs
write.csv(resdata.annot_pCO2History.LowModerate, "Output/Transcriptomics/DESeq2/pCO2History_Low_v_Moderate.csv") # write



# ========================================================== 
#  CO2 Cohorts - Moderate v. High   
# ==========================================================
# Grab DEGs from  the 'dds.run' High
DEGresults_pCO2History.ModerateHigh         <- results(dds_2filt, 
                                                  contrast=c("pCO2_history", 
                                                             "moderate", "severe"),
                                                  alpha = 0.05) # FDR is the alpha 0.05

numDEGs.pval_pCO2History.ModerateHigh       <- data.frame(table(
                                                      DEGresults_pCO2History.ModerateHigh$pvalue<0.05)
                                                     )[2,2] # DEGs - NOT considering LFC - just p adj


numDEGs.padj_pCO2History.ModerateHigh       <- data.frame(table(
                                               DEGresults_pCO2History.ModerateHigh$padj<0.05))[2,2] 


DEGresults.ordered_pCO2History.ModerateHigh <- DEGresults_pCO2History.ModerateHigh[order(
                                                  DEGresults_pCO2History.ModerateHigh$padj), ] # Order by adjusted p-value

num.UpReg_pCO2History.ModerateHigh          <- sum((DEGresults.ordered_pCO2History.ModerateHigh$log2FoldChange
                                               [1:numDEGs.padj_pCO2History.ModerateHigh] > 0) == TRUE) #  LFC >= 1

num.DownReg_pCO2History.ModerateHigh        <- sum((DEGresults.ordered_pCO2History.ModerateHigh$log2FoldChange
                                               [1:numDEGs.padj_pCO2History.ModerateHigh] < 0) == TRUE) # LFC >= 1

total_pCO2History.ModerateHigh              <- sum(num.UpReg_pCO2History.ModerateHigh,num.DownReg_pCO2History.ModerateHigh) # sum of DEGs with the criteria pdj < 0.05 + LFC>1 (< -1)
# Write results - covert to as.data.frame for the ordered results
resdata_pCO2History.ModerateHigh            <- merge(as.data.frame(DEGresults.ordered_pCO2History.ModerateHigh), 
                                                as.data.frame(counts(dds_2filt, normalized=TRUE)), 
                                                by="row.names", 
                                                sort=FALSE) %>% dplyr::filter(padj < 0.05)## Merge with normalized count data
names(resdata_pCO2History.ModerateHigh)[1] <- "Airradians_TranscriptID"
resdata.annot_pCO2History.ModerateHigh     <- merge(Airr_Cvirg_annotation,resdata_pCO2History.ModerateHigh, 
                                               by = 'Airradians_TranscriptID') 
# View(resdata.annot_pCO2History.ModerateHigh)
nrow(resdata.annot_pCO2History.ModerateHigh) == total_pCO2History.ModerateHigh
resdata.annot_pCO2History.ModerateHigh     <- resdata.annot_pCO2History.ModerateHigh[order(
                                             resdata.annot_pCO2History.ModerateHigh$padj),]  # Order by adjusted p-value
# write out annotated DESeq2 DEGs
write.csv(resdata.annot_pCO2History.ModerateHigh, "Output/Transcriptomics/DESeq2/pCO2History_Moderate_v_High.csv") # write

```




### VENN Diagram
```{r}
library(dplyr)
library(VennDiagram)
library("ggVennDiagram")
library(ggvenn)
library(gridExtra)

DEGs_LowHigh      <- resdata.annot_pCO2History.LowHigh %>% 
                                    dplyr::select(Airradians_TranscriptID, 
                                                  log2FoldChange,blastxEval_CvirgGeneID) %>%
                                    dplyr::mutate(Dir = case_when(log2FoldChange > 0 ~ "up",
                                                                  log2FoldChange < 0 ~ "down"),
                                                  About = "Low_v_High")

DEGs_LowModerate  <- resdata.annot_pCO2History.LowModerate %>% 
                                    dplyr::select(Airradians_TranscriptID, 
                                                  log2FoldChange,blastxEval_CvirgGeneID) %>%
                                    dplyr::mutate(Dir = case_when(log2FoldChange > 0 ~ "up",
                                                                  log2FoldChange < 0 ~ "down"),
                                                  About = "Low_v_Moderate")

DEGs_ModerateHigh <- resdata.annot_pCO2History.ModerateHigh %>% 
                                    dplyr::select(Airradians_TranscriptID, 
                                                  log2FoldChange,blastxEval_CvirgGeneID) %>%
                                    dplyr::mutate(Dir = case_when(log2FoldChange > 0 ~ "up",
                                                                  log2FoldChange < 0 ~ "down"),
                                                  About = "Moderate_v_High")

# Merge
bind1       <- rbind(DEGs_LowHigh, DEGs_LowModerate)
Master_DEGs <- rbind(bind1, DEGs_ModerateHigh)




Low_cohort_master <- list(
  vHigh.UP      = (Master_DEGs %>% dplyr::filter(About %in% 'Low_v_High' & 
                                                  Dir %in% 'up'))$Airradians_TranscriptID, 
  vMod.UP      = (Master_DEGs %>% dplyr::filter(About %in% 'Low_v_Moderate' & 
                                                  Dir %in% 'up'))$Airradians_TranscriptID,
  vHigh.DOWN      = (Master_DEGs %>% dplyr::filter(About %in% 'Low_v_High' & 
                                                  Dir %in% 'down'))$Airradians_TranscriptID, 
  vMod.DOWN      = (Master_DEGs %>% dplyr::filter(About %in% 'Low_v_Moderate' & 
                                                  Dir %in% 'down'))$Airradians_TranscriptID
)


Mod_cohort_master <- list(
  vLow.UP      = (Master_DEGs %>% dplyr::filter(About %in% 'Low_v_Moderate' & 
                                                  Dir %in% 'down'))$Airradians_TranscriptID, 
  vHigh.UP     = (Master_DEGs %>% dplyr::filter(About %in% 'Moderate_v_High' & 
                                                  Dir %in% 'up'))$Airradians_TranscriptID,
  vLow.DOWN    = (Master_DEGs %>% dplyr::filter(About %in% 'Low_v_Moderate' & 
                                                 Dir %in% 'up'))$Airradians_TranscriptID, 
  vHigh.DOWN   = (Master_DEGs %>% dplyr::filter(About %in% 'Moderate_v_High' & 
                                                  Dir %in% 'down'))$Airradians_TranscriptID
)


High_cohort_master <- list(
  vLow.UP      = (Master_DEGs %>% dplyr::filter(About %in% 'Low_v_High' & 
                                                  Dir %in% 'down'))$Airradians_TranscriptID, 
  vMod.UP      = (Master_DEGs %>% dplyr::filter(About %in% 'Moderate_v_High' & 
                                                  Dir %in% 'down'))$Airradians_TranscriptID,
  vLow.DOWN      = (Master_DEGs %>% dplyr::filter(About %in% 'Low_v_High' & 
                                                  Dir %in% 'up'))$Airradians_TranscriptID, 
  vMod.DOWN      = (Master_DEGs %>% dplyr::filter(About %in% 'Moderate_v_High' & 
                                                  Dir %in% 'up'))$Airradians_TranscriptID
)


All <- list(
  `H>M`      = (Master_DEGs %>% dplyr::filter(About %in% 'Moderate_v_High' & 
                                                  Dir %in% 'down'))$Airradians_TranscriptID,
  `H>L`      = (Master_DEGs %>% dplyr::filter(About %in% 'Low_v_High' & 
                                                  Dir %in% 'down'))$Airradians_TranscriptID, 
  `M>L`      = (Master_DEGs %>% dplyr::filter(About %in% 'Low_v_Moderate' & 
                                                  Dir %in% 'down'))$Airradians_TranscriptID,
  `M>H`      = (Master_DEGs %>% dplyr::filter(About %in% 'Moderate_v_High' & 
                                                  Dir %in% 'up'))$Airradians_TranscriptID,
  `L>H`      = (Master_DEGs %>% dplyr::filter(About %in% 'Low_v_High' & 
                                                  Dir %in% 'up'))$Airradians_TranscriptID, 
  `L>M`      = (Master_DEGs %>% dplyr::filter(About %in% 'Low_v_Moderate' & 
                                                  Dir %in% 'up'))$Airradians_TranscriptID
)

pdf("Output/Transcriptomics/DESeq2/Cohorts_DEGs_Venn.pdf", width=6, height=6)
print(All_Venn)
dev.off()

library(ggVennDiagram)
All_Venn            <- ggVennDiagram(All, 
                                        color = 1, 
                                        lwd = 0.7, 
                                        label = "count")  + 
                                  scale_fill_gradient(low = "white", high = "#4981BF")  + 
                                  ggtitle("Low pCO2 Cohort response to Elevated pCO2") +
                                  theme(legend.position = "none")

# 64 genes that low downregulated relative to moderate and high (necessary under elevated pCO2?)
# motor proteins (unconventional myosin-XV, dynein gamma chain, flagellar outer arm), ubiquitination oproteasomal degredation
# 
LowpCO2_cohort_Down <- Master_DEGs %>%  
                           dplyr::filter(About %in% c('Low_v_High','Low_v_Moderate') &
                                         Dir %in% 'down') %>% 
                       dplyr::mutate(log2FoldChange = abs(log2FoldChange)) %>% 
                       dplyr::arrange(desc(log2FoldChange)) %>% 
                       dplyr::filter(!blastxEval_CvirgGeneID %in% 'uncharacterized') %>% 
                       na.omit() %>% 
                       group_by(Airradians_TranscriptID) %>%
                       add_count(name = "id_occurrence") %>% 
                       dplyr::filter(id_occurrence == 2) %>% 
                       dplyr::select(c(Airradians_TranscriptID, blastxEval_CvirgGeneID, log2FoldChange))

LowpCO2_cohort_Down <- setDT(LowpCO2_cohort_Down)[, list(log2FoldChange=if(.N==2) median(log2FoldChange) else min(log2FoldChange)) , 
                           by = (blastxEval_CvirgGeneID)] %>% 
                           dplyr::rename(MEDIANlog2FoldChange = log2FoldChange) %>% 
                           dplyr::arrange(desc(MEDIANlog2FoldChange))

write.csv(LowpCO2_cohort_Down, "Output/Transcriptomics/DESeq2/Venn_intersections/Low_expression_LowpCO2.csv")
# 6 genes that Moderate expressed higher than low and high 
# - ceruloplasmin-like, far upstream element-binding protein 1-like, multidrug resistance protein 1-like
ModpCO2_cohort_Up_1 <- Master_DEGs %>% dplyr::filter(About %in% 'Low_v_Moderate' & Dir %in% 'down')
ModpCO2_cohort_Up_2 <- Master_DEGs %>% dplyr::filter(About %in% 'Moderate_v_High' & Dir %in% 'up')
ModpCO2_cohort_Up   <- rbind(ModpCO2_cohort_Up_1, ModpCO2_cohort_Up_2) %>% 
                       dplyr::mutate(log2FoldChange = abs(log2FoldChange)) %>% 
                       dplyr::arrange(desc(log2FoldChange)) %>% 
                       dplyr::filter(!blastxEval_CvirgGeneID %in% 'uncharacterized') %>% 
                       na.omit() %>% 
                       group_by(Airradians_TranscriptID) %>%
                       add_count(name = "id_occurrence") %>% 
                       dplyr::filter(id_occurrence == 2) %>% 
                       dplyr::select(c(Airradians_TranscriptID, blastxEval_CvirgGeneID, log2FoldChange))

ModpCO2_cohort_Up <- setDT(ModpCO2_cohort_Up)[, list(log2FoldChange=if(.N==2) median(log2FoldChange) else min(log2FoldChange)) , 
                           by = (blastxEval_CvirgGeneID)] %>% 
                           dplyr::rename(MEDIANlog2FoldChange = log2FoldChange) %>% 
                           dplyr::arrange(desc(MEDIANlog2FoldChange))
  
write.csv(ModpCO2_cohort_Up, "Output/Transcriptomics/DESeq2/Venn_intersections/High_expression_ModeratepCO2.csv")

# 4 genes that High had downregulated relative to low and moderate 
# - cytochrome P450 2C8-like
HighpCO2_cohort_DOWN <- Master_DEGs %>% dplyr::filter(About %in% c('Low_v_High','Moderate_v_High') & 
                                                          Dir %in% 'up') %>% 
                       dplyr::mutate(log2FoldChange = abs(log2FoldChange)) %>% 
                       dplyr::arrange(desc(log2FoldChange)) %>% 
                       dplyr::filter(!blastxEval_CvirgGeneID %in% 'uncharacterized') %>% 
                       na.omit() %>% 
                       group_by(Airradians_TranscriptID) %>%
                       add_count(name = "id_occurrence") %>% 
                       dplyr::filter(id_occurrence == 2) %>% 
                       dplyr::select(c(Airradians_TranscriptID, blastxEval_CvirgGeneID, log2FoldChange))

HighpCO2_cohort_DOWN <- setDT(HighpCO2_cohort_DOWN)[, list(log2FoldChange=if(.N==2) median(log2FoldChange) else min(log2FoldChange)) , 
                           by = (blastxEval_CvirgGeneID)] %>% 
                           dplyr::rename(MEDIANlog2FoldChange = log2FoldChange) %>% 
                           dplyr::arrange(desc(MEDIANlog2FoldChange))

write.csv(HighpCO2_cohort_DOWN, "Output/Transcriptomics/DESeq2/Venn_intersections/Low_expression_HighpCO2.csv")
# 5 genes that moderate downregulated relative to to low and high
# - WD repeat-containing protein 6-like, biogenesis of lysosome-related organelles complex 1 subunit 1-like
ModpCO2_cohort_Down_1 <- Master_DEGs %>% dplyr::filter(About %in% 'Low_v_Moderate' & Dir %in% 'up')
ModpCO2_cohort_Down_2 <- Master_DEGs %>% dplyr::filter(About %in% 'Moderate_v_High' & Dir %in% 'down')
ModpCO2_cohort_Down   <- rbind(ModpCO2_cohort_Down_1, ModpCO2_cohort_Down_2) %>% 
                       dplyr::mutate(log2FoldChange = abs(log2FoldChange)) %>% 
                       dplyr::arrange(desc(log2FoldChange)) %>% 
                       dplyr::filter(!blastxEval_CvirgGeneID %in% 'uncharacterized') %>% 
                       na.omit() %>% 
                       group_by(Airradians_TranscriptID) %>%
                       add_count(name = "id_occurrence") %>% 
                       dplyr::filter(id_occurrence == 2) %>% 
                       dplyr::select(c(Airradians_TranscriptID, blastxEval_CvirgGeneID, log2FoldChange))
  
ModpCO2_cohort_Down <- setDT(ModpCO2_cohort_Down)[, list(log2FoldChange=if(.N==2) median(log2FoldChange) else min(log2FoldChange)) , 
                           by = (blastxEval_CvirgGeneID)] %>% 
                           dplyr::rename(MEDIANlog2FoldChange = log2FoldChange) %>% 
                           dplyr::arrange(desc(MEDIANlog2FoldChange))
write.csv(ModpCO2_cohort_Down, "Output/Transcriptomics/DESeq2/Venn_intersections/Low_expression_ModeratepCO2.csv")


# 4 genes that High had downregulated relative to low and moderate 
# - cytochrome P450 2C8-like
HighpCO2_cohort_UP <- Master_DEGs %>% dplyr::filter(About %in% c('Low_v_High','Moderate_v_High') & 
                                                          Dir %in% 'down') %>% 
                       dplyr::mutate(log2FoldChange = abs(log2FoldChange)) %>% 
                       dplyr::arrange(desc(log2FoldChange)) %>% 
                       dplyr::filter(!blastxEval_CvirgGeneID %in% 'uncharacterized') %>% 
                       na.omit() %>% 
                       group_by(Airradians_TranscriptID) %>%
                       add_count(name = "id_occurrence") %>% 
                       dplyr::filter(id_occurrence == 2) %>% 
                       dplyr::select(c(Airradians_TranscriptID, blastxEval_CvirgGeneID, log2FoldChange))

HighpCO2_cohort_UP <- setDT(HighpCO2_cohort_UP)[, list(log2FoldChange=if(.N==2) median(log2FoldChange) else min(log2FoldChange)) , 
                           by = (blastxEval_CvirgGeneID)] %>% 
                           dplyr::rename(MEDIANlog2FoldChange = log2FoldChange) %>% 
                           dplyr::arrange(desc(MEDIANlog2FoldChange))

write.csv(HighpCO2_cohort_UP, "Output/Transcriptomics/DESeq2/Venn_intersections/High_expression_HighpCO2.csv")
```

## Plot the DEG statistics

```{r histogram}
# histogram of FDR
png("../../Output/Transcriptomics/DESeq2/F1_juveniles/F1_FDR_hist.png", 1000, 1000, pointsize=20)
hist(DEGresults_F1$padj, breaks=20, col="grey") # view histogram
abline(h=c( (nrow(DEGresults_F1)*0.05), 
            ((table(DEGresults_F1$padj < 0.1)[2]) + (nrow(DEGresults_F1)*0.1)),
            ((table(DEGresults_F1$padj < 0.05)[2]) + (nrow(DEGresults_F1)*0.05)) ),
                  col=c("blue", "red", "red"), lty=c(1,2,1), lwd=c(1, 3, 1)) # add line at 
dev.off()

png("../../Output/Transcriptomics/DESeq2/F2_juveniles/F2_FDR_hist.png", 1000, 1000, pointsize=20)
hist(DEGresults_F2$padj, breaks=20, col="grey") # view histogram
abline(h=c( (nrow(DEGresults_F2)*0.05), 
            ((table(DEGresults_F2$padj < 0.1)[2]) + (nrow(DEGresults_F2)*0.1)),
            ((table(DEGresults_F2$padj < 0.05)[2]) + (nrow(DEGresults_F2)*0.05)) ),
                  col=c("blue", "red", "red"), lty=c(1,2,1), lwd=c(1, 3, 1)) # add line at 
dev.off()

?EnhancedVolcano
# volcano plot 
# png("../../Output/Transcriptomics/DESeq2/F1_juveniles/F1_DEGs-VolcanoPlot.png", 1000, 1000, pointsize=20)
pdf("../../Output/Transcriptomics/DESeq2/F1_juveniles/F1_DEGs-VolcanoPlot.pdf", width = 7, height = 5)
EnhancedVolcano(DEGresults_F1,
                #lab = rownames(DEGresults_F2),
                lab = NA, # no labels
                #labSize = 4, # remove gene labels
                x = 'log2FoldChange',
                y = 'padj',
                title = 'F1s: pCO2 Treatment (Low v High)',
                subtitle = "DESeq2 - Differential expression",
                FCcutoff = 0.5,
                pCutoff = 0.05,
                 ylim = c(0,5),
                pointSize = 4.0,
                labSize = 3,
                colAlpha = 1,
                legendPosition = 'right',
                legendLabSize = 12,
                legendIconSize = 2,
                widthConnectors = 0.75)
dev.off()


# png("../../Output/Transcriptomics/DESeq2/F2_juveniles/F2_DEGs-VolcanoPlot.png", 1000, 1000, pointsize=20)
pdf("../../Output/Transcriptomics/DESeq2/F2_juveniles/F2_DEGs-VolcanoPlot.pdf", width = 7, height = 5)
EnhancedVolcano(DEGresults_F2,
                #lab = rownames(DEGresults_F2),
                lab = NA, # no labels
                #labSize = 4, # remove gene labels
                x = 'log2FoldChange',
                y = 'padj',
                title = 'F2s: pCO2 Treatment (Low v High)',
                subtitle = "DESeq2 - Differential expression",
                FCcutoff = 0.5,
                 ylim = c(0,5),
                pCutoff = 0.05,
                pointSize = 4.0,
                colAlpha = 1,
                legendPosition = 'right',
                legendLabSize = 12,
                legendIconSize = 4.0,
                widthConnectors = 0.75)
dev.off()

# Plot dispersion 

png("../../Output/Transcriptomics/DESeq2/F1_juveniles/F1_dispersions.png", 1000, 1000, pointsize=20)
plotDispEsts(dds.run_F1, main="dispersions")
dev.off()

png("../../Output/Transcriptomics/DESeq2/F2_juveniles/F2_dispersions.png", 1000, 1000, pointsize=20)
plotDispEsts(dds.run_F2, main="dispersions")
dev.off()


```

## Figures for F1 and F2 data 
```{r, Figs}

PCA_F1_vst <- ggplot(pcaData_vstExp_F1, 
                     aes(x = PC1, 
                         y = PC2, 
                         color = pCO2, 
                         label=name)) +
                geom_point(size =4) +
                #scale_fill_manual(c('grey50', 'white')) +
                theme_classic() +
                stat_ellipse() +
                ylim(-20,20) +
                xlim(-35,35) +
                theme(legend.position = 'none',axis.text = element_text(size = 12)) +
                ggtitle("F1 PCA: vst expression data") +
                xlab(paste0("PC1: ", percentVar_vstExp_F1[1], "% variance")) + 
                ylab(paste0("PC2: ", percentVar_vstExp_F1[2], "% variance")) +
                coord_fixed()
PCA_F2_vst <- ggplot(pcaData_vstExp_F2, 
                     aes(x = PC1, 
                         y = PC2, 
                         color = pCO2, 
                         label=name)) +
                geom_point(size =4) +
                theme_classic() +
                stat_ellipse() +
                ylim(-20,20) +
                xlim(-35,35) +
                theme(legend.position = 'none',axis.text = element_text(size = 12)) +
                ggtitle("F2 PCA: vst expression data") +
                xlab(paste0("PC1: ", percentVar_vstExp_F2[1], "% variance")) + 
                ylab(paste0("PC2: ", percentVar_vstExp_F2[2], "% variance")) +
                coord_fixed()

F1_Volcano <- EnhancedVolcano(DEGresults_F1,
                #lab = rownames(DEGresults_F2),
                lab = NA, # no labels
                #labSize = 4, # remove gene labels
                x = 'log2FoldChange',
                y = 'padj',
                # title = 'F1s: pCO2 Treatment (Low v High)',
                # subtitle = "DESeq2 - Differential expression",
                titleLabSize = NA,
                FCcutoff = 0.5,
                pCutoff = 0.05,
                pointSize = 2,
                axisLabSize = 10,
                labSize = 3,
                colAlpha = 1,
                legendPosition = "none",
                # legendLabSize = 12,
                # legendIconSize = 2,
                widthConnectors = 0.75)
F2_Volcano <- EnhancedVolcano(DEGresults_F2,
                #lab = rownames(DEGresults_F2),
                lab = NA, # no labels
                #labSize = 4, # remove gene labels
                x = 'log2FoldChange',
                y = 'padj',
                # title = 'F2s: pCO2 Treatment (Low v High)',
                # subtitle = "DESeq2 - Differential expression",
                # title = NA,
                titleLabSize = NA,
                FCcutoff = 0.5,
                pCutoff = 0.05,
                pointSize = 2,
                axisLabSize = 10,
                labSize = 3,
                colAlpha = 1,
                legendPosition = "none",
                # legendLabSize = 12,
                widthConnectors = 0.75)
# ?EnhancedVolcano
ggarrange(PCA_F1_vst, PCA_F2_vst, 
          F1_Volcano, F2_Volcano,
          nrow=2, ncol=2)

pdf("../../Output/Transcriptomics/DESeq2/PCA_F1_F2_vst.pdf", width = 14, height = 5)
ggarrange(PCA_F1_vst, PCA_F2_vst, 
          nrow=1, ncol=2)
dev.off()
```

## plot heatmap

```{r heatmap for transformed data}

# Plot pheatmap map rlog
# setup the save pheatmap function
save_pheatmap <- function(x, filename, width=1000, height=960) { # template for saving pheatmap outputs
  stopifnot(!missing(x))
  stopifnot(!missing(filename))
  png(filename,width = width, height=height)
  grid::grid.newpage()
  grid::grid.draw(x$gtable)
  dev.off()
}
# colors for the heatmap
# colorblind palette
colorBlindBlack8  <- c("#000000", "#E69F00", "#56B4E9", "#009E73", 
                         "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
pie(rep(1, 8), col = colorBlindBlack8) # view the pie chart for colorblindness
ann_colors = list(
    pCO2 = c(Low="#56B4E9", High="#E69F00"))
df_annot.col         <- as.data.frame(colData(dds.run_F1)[c("pCO2")])


# rlogExp
rlogExp_F1 <- rlogTransformation(dds.run_F1)
# use topgenes.IDs from the chunk above - this calls only the pdj > 0.05 from the output results of 'dds.run'
topgenes_rlog.counts_F1      <- assay(rlogExp_F1)[topgenes.IDs_F1,] 
topgenes_rlog.corrected_F1   <- topgenes_rlog.counts_F1 - rowMeans(topgenes_rlog.counts_F1) # subtract from the row mean to get +/- 0 to normalize and ease the aesthetic
df_annot.col_F1         <- as.data.frame(colData(dds.run_F1)[c("pCO2")])
rlog.heatmap <- pheatmap(topgenes_rlog.corrected_F1, 
                            annotation_col=df_annot.col_F1, 
                            main = "Ambient versus Moderate (169 total DEGs)",
                            cutree_cols = 2,
                            cutree_rows = 2,
                            annotation_legend = TRUE,
                            annotation_colors = ann_colors,
                            show_rownames = TRUE, # Boolean for gene names
                            labels_col=df_annot.col_F1$pCO2, 
                            angle_col = 0,
                            fontsize = 8,
                            legend = TRUE)
save_pheatmap(rlog.heatmap, filename = "../../Output/Transcriptomics/DESeq2/F1_juveniles/F1_rlog_heatmap.png") #Save pheatmap



# vstExp
# use topgenes.IDs from the chunk above - this calls only the pdj > 0.05 from the output results of 'dds.run'
vstExp_F1 <- vst(dds.run_F1)
topgenes_vst.counts_F1      <- assay(vstExp_F1)[topgenes.IDs_F1,] 
topgenes_vst.corrected_F1   <- topgenes_vst.counts_F1 - rowMeans(topgenes_vst.counts_F1) # subtract from the row mean to get +/- 0 to normalize and ease the aesthetic
vst.heatmap_F1 <- pheatmap(topgenes_vst.corrected_F1, 
                            annotation_col=df_annot.col_F1, 
                            main = "Ambient versus Moderate (169 total DEGs)",
                            cutree_cols = 3,
                            cutree_rows = 2,
                            annotation_legend = TRUE,
                            annotation_colors = ann_colors,
                            show_rownames = TRUE, # Boolean for gene names
                            labels_col=df_annot.col_F1$pCO2, 
                            angle_col = 0,
                            fontsize = 8,
                            legend = TRUE)
save_pheatmap(vst.heatmap_F1, filename = "../../Output/Transcriptomics/DESeq2/F1_juveniles/vst_heatmap.png") #Save pheatmap

```


# plot all the DEGs
```{r plot the rlogExp}

# ========================================================== 
#  F1s  
# ==========================================================
rlogExp_F1_df        <- as.data.frame(assay(rlogExp_F1)) %>% 
                           tibble::rownames_to_column("Gene.ID")
rlogExp_F1_DEGs      <- rlogExp_F1_df %>% 
                          dplyr::filter(Gene.ID %in% topgenes.IDs_F1)
rlogExp_F1_DEGs_melt <- reshape2::melt(rlogExp_F1_DEGs, id.vars='Gene.ID') %>% 
                        dplyr::rename(Sample.Name = variable, # to merge with 'Exp.metadata'
                                      rlogExp = value) # the data to get mean SE values 
DEGs_annotated_F1    <- Airr_Cvirg_annotation %>% 
                          dplyr::filter(Airradians_TranscriptID %in% topgenes.IDs_F1) %>% 
                          dplyr::select(Airradians_TranscriptID,blastxEval_CvirgProteinID) %>% 
                          dplyr::rename(Gene.ID = Airradians_TranscriptID,
                                      Protein.ID = blastxEval_CvirgProteinID)

rlogExp_F1_DEGs_annotated <- merge(DEGs_annotated_F1, rlogExp_F1_DEGs_melt, by = "Gene.ID")
rlogExp_F1_DEGs_MASTER    <- merge(rlogExp_F1_DEGs_annotated,Exp.metadata, by = "Sample.Name")
# library(stringr)
rlogExp_F1_DEGs_MEAN.SE   <- rlogExp_F1_DEGs_MASTER %>% 
                              dplyr:::group_by(Gene.ID, Protein.ID,pCO2) %>% 
                              dplyr::summarise(mean.rlogExp    = mean(rlogExp),
                                               sd.rlogExp     = sd(rlogExp),
                                               n = n(),
                                               sderror.rlogExp = sd.rlogExp/(sqrt(n))) %>% 
                              dplyr::select(-n) %>% 
                              dplyr::mutate(Protein_contig = paste(substr(Protein.ID,1,30),
                                                                   str_split(Gene.ID, "\\.", simplify=T)[,3],
                                                                   sep="_"))
pd <- position_dodge(0.3)
for(i in 1:length(topgenes.IDs_F1)) {
  
  geneID_loop  <- topgenes.IDs_F1[i]
  
  df_loop      <- rlogExp_F1_DEGs_MEAN.SE %>% 
                  dplyr::filter(Gene.ID %in% geneID_loop) %>% 
                  dplyr::mutate(Protein_contig = str_remove_all(Protein_contig,"/"))
  
  plot_loop    <- df_loop %>% 
                    ggplot(aes(x=pCO2, 
                               y=mean.rlogExp, 
                               fill=pCO2)) +  # , colour=supp, group=supp))
                    theme_classic() +
                    geom_errorbar(aes(ymin=mean.rlogExp-sderror.rlogExp, 
                                      ymax=mean.rlogExp+sderror.rlogExp), 
                                  colour="black", 
                                  width=.1,
                                  position=pd) +
                    geom_point(size = 4, 
                               shape=21,
                               position=pd) +            
                    xlab("pCO2 treatment") +
                    ylab("") +                 # note the mean was first by sample ID THEN by treatment
                    scale_fill_manual(values=c("white",
                                               "grey50")) +
                    ggtitle(df_loop[1,7]) +
                    theme(axis.text.x = element_text(size = 20),
                          axis.text.y = element_text(size = 20),
                          axis.ticks.length=unit(.25, "cm")) +
                    theme(legend.position = "none")
  
if(df_loop[1,4]>df_loop[2,4]) { # Upregulated, high pCO2 == high expression
  pdf(paste("../../Output/Transcriptomics/DESeq2/F1_juveniles/Plots_MeanSE/Upregulated_DEGs/",df_loop[1,7],".pdf", sep = ''), width=5, height=5)
  print(plot_loop)
  dev.off() 
  } else {
  pdf(paste("../../Output/Transcriptomics/DESeq2/F1_juveniles/Plots_MeanSE/Downregulated_DEGs/",df_loop[1,7],".pdf", sep = ''), width=5, height=5)
  print(plot_loop)
  dev.off()  
  }
}




# ========================================================== 
#  F2s  
# ==========================================================
rlogExp_F2_df        <- as.data.frame(assay(rlogExp_F2)) %>% 
                           tibble::rownames_to_column("Gene.ID")
rlogExp_F2_DEGs      <- rlogExp_F2_df %>% 
                          dplyr::filter(Gene.ID %in% topgenes.IDs_F2)
rlogExp_F2_DEGs_melt <- reshape2::melt(rlogExp_F2_DEGs, id.vars='Gene.ID') %>% 
                        dplyr::rename(Sample.Name = variable, # to merge with 'Exp.metadata'
                                      rlogExp = value) # the data to get mean SE values 
DEGs_annotated_F2    <- Airr_Cvirg_annotation %>% 
                          dplyr::filter(Airradians_TranscriptID %in% topgenes.IDs_F2) %>% 
                          dplyr::select(Airradians_TranscriptID,blastxEval_CvirgProteinID) %>% 
                          dplyr::rename(Gene.ID = Airradians_TranscriptID,
                                      Protein.ID = blastxEval_CvirgProteinID)

rlogExp_F2_DEGs_annotated <- merge(DEGs_annotated_F2, rlogExp_F2_DEGs_melt, by = "Gene.ID")
rlogExp_F2_DEGs_MASTER    <- merge(rlogExp_F2_DEGs_annotated,Exp.metadata, by = "Sample.Name")
# library(stringr)
rlogExp_F2_DEGs_MEAN.SE   <- rlogExp_F2_DEGs_MASTER %>% 
                              dplyr:::group_by(Gene.ID, Protein.ID,pCO2) %>% 
                              dplyr::summarise(mean.rlogExp    = mean(rlogExp),
                                               sd.rlogExp     = sd(rlogExp),
                                               n = n(),
                                               sderror.rlogExp = sd.rlogExp/(sqrt(n))) %>% 
                              dplyr::select(-n) %>% 
                              dplyr::mutate(Protein_contig = paste(substr(Protein.ID,1,30),
                                                                   str_split(Gene.ID, "\\.", simplify=T)[,3],
                                                                   sep="_"))
pd <- position_dodge(0.3)
for(i in 1:length(topgenes.IDs_F2)) {
  
  geneID_loop  <- topgenes.IDs_F2[i]
  
  df_loop      <- rlogExp_F2_DEGs_MEAN.SE %>% 
                  dplyr::filter(Gene.ID %in% geneID_loop) %>% 
                  dplyr::mutate(Protein_contig = str_remove_all(Protein_contig,"/"))
  
  plot_loop    <- df_loop %>% 
                    ggplot(aes(x=pCO2, 
                               y=mean.rlogExp, 
                               fill=pCO2)) +  # , colour=supp, group=supp))
                    theme_classic() +
                    geom_errorbar(aes(ymin=mean.rlogExp-sderror.rlogExp, 
                                      ymax=mean.rlogExp+sderror.rlogExp), 
                                  colour="black", 
                                  width=.1,
                                  position=pd) +
                    geom_point(size = 4, 
                               shape=21,
                               position=pd) +            
                    xlab("pCO2 treatment") +
                    ylab("") +                 # note the mean was first by sample ID THEN by treatment
                    scale_fill_manual(values=c("white",
                                               "grey50")) +
                    ggtitle(df_loop[1,7]) +
                    theme(axis.text.x = element_text(size = 20),
                          axis.text.y = element_text(size = 20),
                          axis.ticks.length=unit(.25, "cm")) +
                    theme(legend.position = "none")
  
if(df_loop[1,4]>df_loop[2,4]) { # Upregulated, high pCO2 == high expression
  pdf(paste("../../Output/Transcriptomics/DESeq2/F2_juveniles/Plots_MeanSE/Upregulated_DEGs/",df_loop[1,7],".pdf", sep = ''), width=5, height=5)
  print(plot_loop)
  dev.off() 
  } else {
  pdf(paste("../../Output/Transcriptomics/DESeq2/F2_juveniles/Plots_MeanSE/Downregulated_DEGs/",df_loop[1,7],".pdf", sep = ''), width=5, height=5)
  print(plot_loop)
  dev.off()  
  }
}
```

