---
title: "DESeq2"
author: "Samuel Gurr"
date: "2024-02-21"
output: pdf_document
---


### Set working directory
```{r setup, include=TRUE}
knitr::opts_chunk$set(echo = TRUE, 
                      warning = FALSE, 
                      message = FALSE, 
                      cache = TRUE)

knitr::opts_knit$set(root.dir = "C:/Users/samjg/Documents/Github_repositories/Airradians_CellularMolecular_OA/RAnalysis")
# knitr::opts_knit$set(root.dir = "C:/Users/samuel.gurr/Documents/Github_repositories/Airradians_CellularMolecular_OA/RAnalysis")
```

### Load libraries
```{r load_libraries, include = TRUE}
# load libraries - notes show the install command needed to install (pre installed)
library(tidyr)
library(DESeq2) # note: this was previously installed with the command `BiocManager::install("DESeq2")`
library(edgeR)
library(goseq)
library(dplyr)
library(GenomicFeatures)
library(data.table)
library(calibrate)
library(affycoretools) # note: this was previously installed with the BiocManager::install("affycoretools")
library(data.table)
library(vsn)
# Plotting
library(ggplot2)
library(cowplot)
library(pheatmap)
library(gplots)
library(RColorBrewer)
library(EnhancedVolcano)  # note: this was previously installed with the command `BiocManager::install("EnhancedVolcano")`
library(tweeDEseq)
library(pcaExplorer) # note: this was previously installed with the command `BiocManager::install("pcaExplorer")
library(ggpubr)
library(stringr)
```

### Sample metadata - Experimental treatments/groups
```{r experiment_data, include = TRUE}
### experiment metadata [from Count_Matrix_Stats.Filter.R]  - convert characaters to factors for DESeq2
exp_metadata <- read.csv(file="Data/Transcriptomics/metadata/metadata.csv", sep=',', header=TRUE) %>% 
                    dplyr::mutate(All_pCO2 = paste0(pCO2_history, pCO2_exposure))

# NOTE: the Smaple_num S32 should be omitted - it entials a single individual from severe pCO2 history 
# exposure to moderate pCO2 - omit from both the expreiment metadata and from the count amtrices for building dds objects! 
exp_metadata <- exp_metadata %>% dplyr::filter(!Sample_num == 'S32')# 20230516	80	14	severe	moderate	C7	S32
nrow(exp_metadata) == 35 # TRUE!!
```

### Gene count data 

* below are the raw and filtered read matrices

* 35 samples as the following 

  - Low pCO2 history (N =15), under low, moderate, and high pCO2 exposure (N = 5 ea)
  
  - Moderate pCO2 history (N = 5), only under *matched* moderate pCO2 exposure
    
  - High pCO2 history (N =15), under low, moderate, and high pCO2 exposure (N = 5 ea)

* two questions to take with WGCNA 

  - (1) Response to the challenge, pin low v high history under all challenges (N = 30) 
  
  - (2) Cohort differences under their *matched* exposure (N = 15)

### Read in the raw and filtered data

- 'Raw' - unfiltered read count matrix, formatted as a dataframe

  * raw_counts = raw unfiltered (N = 35 samples)
  
- 'filtered' - using edgeR and CPM thresholds to target mean 1000 reads per gene

  * filtered_counts_all = N = 35; 5 cpm used to acquire mean 1000 reads per gene in 50% samples
  
  * filtered_counts_challenge = N = 30; 5 cpm used to acquire mean 1000 reads per gene in 50% samples
  
  * filtered_counts_cohort = N = 15; 9 cpm used to acquire mean 1000 reads per gene in 33.33% samples
```{r call the count matrices, include = TRUE}
# non-filtered read matrix
raw_counts     <-
                  read.csv(file="Output/Transcriptomics/raw_count_matrix_all.csv", 
                                                               sep=',', 
                                                               header=TRUE)  %>% 
                      dplyr::rename(transcript_id = X) %>% 
                      tibble::column_to_rownames(var = "transcript_id") %>% 
                      dplyr::select(!S32) %>% 
                      as.matrix()  
ncol(raw_counts) # 35 - good! 
nrow(raw_counts) # 49900 - all genes
# gather the 3CPM filtered read matrices 

# all filtered
filtered_counts_all <-
                  read.csv(file="Output/Transcriptomics/Filtered_count_matrix/filtered_count_matrix_all.csv", 
                                                               sep=',', 
                                                               header=TRUE)  %>% 
                      dplyr::rename(transcript_id = X) %>% 
                      tibble::column_to_rownames(var = "transcript_id") %>% 
                      dplyr::select(!S32) %>% 
                      as.matrix()  
ncol(filtered_counts_all) # 35, good! 
nrow(filtered_counts_all) # 10001 subset from the 5CPM cutoff


# challenge
filtered_counts_challenge <-
                  read.csv(file="Output/Transcriptomics/Filtered_count_matrix/filtered_count_matrix_challenge.csv", 
                                                               sep=',', 
                                                               header=TRUE)  %>% 
                      dplyr::rename(transcript_id = X) %>% 
                      tibble::column_to_rownames(var = "transcript_id") %>% 
                      as.matrix()  
ncol(filtered_counts_challenge) # 30, good! 
nrow(filtered_counts_challenge) # 9866 subset from the 5CPM cutoff

# cohort
filtered_counts_cohort <-
                  read.csv(file="Output/Transcriptomics/Filtered_count_matrix/filtered_count_matrix_cohort.csv", 
                                                               sep=',', 
                                                               header=TRUE)  %>% 
                      dplyr::rename(transcript_id = X) %>% 
                      tibble::column_to_rownames(var = "transcript_id") %>% 
                      as.matrix()  
ncol(filtered_counts_cohort) # 15, good! 
nrow(filtered_counts_cohort) # 12163 subset from the 9CPM cutoff

```




```{r filtered cohort data boxplots visual}
filtered_counts_cohor_df <- (as.data.frame((filtered_counts_cohort)))
str(filtered_counts_cohor_df, strict.width = "wrap")
data_st                  <- as.data.frame(t(apply(as.matrix(filtered_counts_cohor_df), 1, scale)))
str(data_st, strict.width = "wrap")
data_st                  <- tibble::rownames_to_column(data_st, "ID")
colnames(data_st)[2:ncol(data_st)] <- colnames(filtered_counts_cohor_df)

nID <- which(is.nan(data_st[,2]))
data_st[nID,2:length(data_st)] <- 0

str(data_st, strict.width = "wrap")

library(reshape2)
library(ggplot2)
ggplot(data = melt(data_st), aes(x=variable, y=value)) + geom_boxplot()

# looks like sample S.33 is poor here - we should ommit
# Note: downsteram without omitting we saw this is an outliter in PCA analysis too 
colnames(data_st)[2:6] <- c('Low.1','Low.2','Low.3','Low.4','Low.5')
colnames(data_st)[7:11] <- c('Mod.1','Mod.2','Mod.3','Mod.4','Mod.5')
colnames(data_st)[12:16] <- c('High.1','High.2','High.3','High.4','High.5')
```

### (1) Response to challenge - read count (raw and filtered) and metadata
* ncol() == 30!
* here we simply call the filtered data '1' for challenge
```{r edit data for #1}
exp_metadata_1    <- exp_metadata %>% dplyr::filter(pCO2_history %in% c('low','severe')) # omit moderate history
raw_counts_1      <- raw_counts[,exp_metadata_1$Sample_num] # omit columns from the raw matrix
filtered_counts_1 <- filtered_counts_challenge # use the filtered data - already catered to the target samples when filtered prior
```

### (2) Cohort differences under *match* - read count (raw and filtered) and metadata
* here we simply call the filtered data '2' for cohort
* omit sample num S33 due to the standardized being off and the downsteram PCA shows it is off
* ncol() == 14!
```{r edit data for #2}
exp_metadata_2    <- exp_metadata %>% dplyr::filter(All_pCO2 %in% c('lowlow', 'moderatemoderate', 'severesevere')) %>% 
                        dplyr::filter(!Sample_num == 'S36')
raw_counts_2      <- raw_counts[,exp_metadata_2$Sample_num]
filtered_counts_2 <- filtered_counts_cohort[,exp_metadata_2$Sample_num] # use the filtered data - already catered to the target samples when filtered prior
ncol(filtered_counts_2)
```


### DESeqDataSet or 'dds' object (using DESeq2) 
```{r build dds object}
# (1) Response to challenge 
dds_1raw <- DESeqDataSetFromMatrix(countData = raw_counts_1,
                                 colData = exp_metadata_1, design = ~pCO2_history+pCO2_exposure) # DESeq Data Set (dds)

dds_1filt <- DESeqDataSetFromMatrix(countData = filtered_counts_1,
                                 colData = exp_metadata_1, design = ~pCO2_history+pCO2_exposure) # DESeq Data Set (dds)

# (2) Cohort differences under *match*
dds_2raw <- DESeqDataSetFromMatrix(countData = raw_counts_2,
                                 colData = exp_metadata_2, design = ~pCO2_history) # DESeq Data Set (dds)

dds_2filt <- DESeqDataSetFromMatrix(countData = filtered_counts_2,
                                 colData = exp_metadata_2, design = ~pCO2_history) # DESeq Data Set (dds)
```


### run DESeq model - Full & Matched

```{r, Run DESeq}
# Full models
# (1) Response to challenge 

dds_1raw       <- DESeq(dds_1raw) # 
dds_1filt      <- DESeq(dds_1filt) # 

# (2) Cohort differences under *match*
dds_2raw     <- DESeq(dds_2raw) # TARGET DATA FOR DESEQ2 COHORT ANLAYSIS
dds_2filt    <- DESeq(dds_2filt) # TRAGET DATA FOR DESEQ2 COHORT ANALYSIS
```

## Transform Exp data 

* Why?
 - the transformed data can be used to run simple PCA 
 - identify whether samples appear as outliers sufficient for omission and rerun dds.run

```{r expression data transformation}

# challenge 
vstExp.raw_challenge       <- vst(dds_1raw)
vstExp.raw_challenge.assay <- assay(vstExp.raw_challenge) # call only the transformed coutns in the dds object
vstExp.raw_challenge.df    <- as.data.frame(vstExp.raw_challenge.assay)
vstExp.raw_challenge.df    <- tibble::rownames_to_column(vstExp.raw_challenge.df, var = "Gene")


write.csv(vstExp.raw_challenge.df, "Output/Transcriptomics/vstExp_raw_count_matrix_challenge.csv")



vstExp.filt_challenge       <- vst(dds_1filt)
vstExp.filt_challenge.assay <- assay(vstExp.filt_challenge) # call only the transformed coutns in the dds object
vstExp.filt_challenge.df    <- as.data.frame(vstExp.filt_challenge.assay)
vstExp.filt_challenge.df    <- tibble::rownames_to_column(vstExp.filt_challenge.df, var = "Gene")


write.csv(vstExp.filt_challenge.df, "Output/Transcriptomics/vstExp_filtered_count_matrix_challenge.csv")



# cohort  
vstExp.raw_cohort       <- vst(dds_2raw)
vstExp.raw_cohort.assay <- assay(vstExp.raw_cohort) # call only the transformed coutns in the dds object
vstExp.raw_cohort.df    <- as.data.frame(vstExp.raw_cohort.assay)
vstExp.raw_cohort.df    <- tibble::rownames_to_column(vstExp.raw_cohort.df, var = "Gene")
vstExp.raw_cohort.df    <- vstExp.raw_cohort.df %>% 
                              mutate_at(c(2:ncol(vstExp.raw_cohort.df)), as.numeric)

write.csv(vstExp.raw_cohort.df, "Output/Transcriptomics/vstExp_raw_count_matrix_cohort.csv")



vstExp.filt_cohort       <- vst(dds_2filt)
vstExp.filt_cohort.assay <- assay(vstExp.filt_cohort) # call only the transformed coutns in the dds object
vstExp.filt_cohort.df    <- as.data.frame(vstExp.filt_cohort.assay)
vstExp.filt_cohort.df    <- tibble::rownames_to_column(vstExp.filt_cohort.df, var = "Gene")
vstExp.filt_cohort.df    <- vstExp.filt_cohort.df %>% 
                              mutate_at(c(2:ncol(vstExp.filt_cohort.df)), as.numeric)

write.csv(vstExp.filt_cohort.df, "Output/Transcriptomics/vstExp_filtered_count_matrix_cohort.csv")

```


## Visualize data using scatterplot matrices
```{r Scatterplot matrix}
# fxns
reg <- function(x, y, ...) {
  points(x,y, ...)
  abline(lm(y~x)) 
  }# made to draw regression line instead of lowess line
panel.cor <- function(x, y, digits = 2, prefix = "", cex.cor, ...)
{
  usr <- par("usr"); on.exit(par(usr))
  par(usr = c(0, 1, 0, 1))
  r <- abs(cor(x, y))
  txt <- format(c(r, 0.123456789), digits = digits)[1]
  txt <- paste0(prefix, txt)
  text(0.5, 0.5, txt, cex = 1.1, font = 4)
}

# now you can run PAirs with correct formatting 

vstExp.filt_cohort.PLOTTING <- vstExp.filt_cohort.df
colnames(vstExp.filt_cohort.PLOTTING)[2:6] <- c('Low.1','Low.2','Low.3','Low.4','Low.5')
colnames(vstExp.filt_cohort.PLOTTING)[7:11] <- c('Mod.1','Mod.2','Mod.3','Mod.4','Mod.5')
colnames(vstExp.filt_cohort.PLOTTING)[12:15] <- c('High.1','High.2','High.3','High.4')

## Low v Moderate  - filtered data pinning 3 from low and 3 from moderate 
pairs(vstExp.filt_cohort.PLOTTING[c(4,5,6,7,9,10)], 
      upper.panel = reg, # replace HERE for panel.smooth #
      cex = 1.5, 
      pch = 19, 
      col = adjustcolor(4, .4), 
      cex.labels = 2, 
      font.labels = 2, 
      lower.panel = panel.cor)
## Moderate v High - filtered data pinning 4 from low and 3 from high 
pairs(vstExp.filt_cohort.PLOTTING[c(7,9,10,12,13,14)], 
                          upper.panel = reg, # replace HERE for panel.smooth #
                          cex = 1.5, pch = 19, 
                          col = adjustcolor(4, .4), 
                          cex.labels = 2, 
                          font.labels = 2, 
                          lower.panel = panel.cor)
## Low v High - filtered data pinning 4 from low and 3 from high 
pairs(vstExp.filt_cohort.PLOTTING[c(4,5,6,12,13,14)], 
                          upper.panel = reg, # replace HERE for panel.smooth #
                          cex = 1.5, pch = 19, 
                          col = adjustcolor(4, .4), 
                          cex.labels = 2, 
                          font.labels = 2, 
                          lower.panel = panel.cor)

```


## Run diagnostic plots
```{r, Diagnostic plotting}
library(ggplot2)
vstExp.filt_cohort.df
hist(vstExp.filt_cohort.df)
meanSdPlot(vstExp.filt_cohort.df)

pcaData_vstExp_2    <- plotPCA(vstExp.filt_cohort, 
                               intgroup = "pCO2_history", 
                               returnData = TRUE)
percentVar_vstExp_2 <- round(100 * attr(pcaData_vstExp_2, "percentVar"))
ggplot(pcaData_vstExp_2, aes(x = PC1, 
                             y = PC2, 
                             color = pCO2_history, 
                             label=name)) +

      geom_point(size =3) +
      theme_classic() +
      stat_ellipse() +
      #theme(text = element_text(size=15)) +
      theme_classic() +
      theme(text = element_text(size=15)) +
      ggtitle("Cohort PCA: vst expression data") +
      xlab(paste0("PC1: ", percentVar_vstExp_2[1], "% variance")) + 
      ylab(paste0("PC2: ", percentVar_vstExp_2[2], "% variance")) +
      coord_fixed()

```


## DEGs - data tables
* Examine DEGs how many? what are they?

### Cohorts data 
```{r}

# Edit the annotation file (Cvirg blasted to annotate draft genome contigs of Airradians)
Airr_Cvirg_annotation <- read.csv(file="Data/Transcriptomics/metadata/seq_id_AirrCvirg_MERGED_master.csv",
                                  sep = ',', 
                                  header = T) %>% 
                        dplyr::select(c('NCBI_refseq', 
                        'Airradians_TranscriptID',
                        'Airradians_GeneID', # Airra IDs
                                 "blastxEval_CvirgTranscriptID",
                                 "blastxEval_CvirgProteinID",
                                 "blastxEval_CvirgGeneID",
                                 "blastxEval_CvirgGOterms")) %>% 
                        dplyr::rename(Airradians_NCBI_refseq = NCBI_refseq,
                                      Airradians_gene_id = Airradians_GeneID)

# ========================================================== 
#  CO2 Cohorts - Low v. High   
# ==========================================================
# Grab DEGs from  the 'dds.run' High
DEGresults_pCO2History.LowHigh         <- results(dds_2filt, 
                                                  contrast=c("pCO2_history", 
                                                             "low", "severe"),
                                                  alpha = 0.05) # FDR is the alpha 0.05

numDEGs.pval_pCO2History.LowHigh       <- data.frame(table(
                                                      DEGresults_pCO2History.LowHigh$pvalue<0.05)
                                                     )[2,2] # DEGs - NOT considering LFC - just p adj


numDEGs.padj_pCO2History.LowHigh       <- data.frame(table(DEGresults_pCO2History.LowHigh$padj<0.05))[2,2] 


DEGresults.ordered_pCO2History.LowHigh <- DEGresults_pCO2History.LowHigh[order(
                                            DEGresults_pCO2History.LowHigh$padj), ] # Order by adjusted p-value

num.UpReg_pCO2History.LowHigh          <- sum((DEGresults.ordered_pCO2History.LowHigh$log2FoldChange
                                               [1:numDEGs.padj_pCO2History.LowHigh] > 1) == TRUE) #  LFC >= 1

num.DownReg_pCO2History.LowHigh        <- sum((DEGresults.ordered_pCO2History.LowHigh$log2FoldChange
                                               [1:numDEGs.padj_pCO2History.LowHigh] < 1) == TRUE) # LFC >= 1

total_pCO2History.LowHigh              <- sum(num.UpReg_pCO2History.LowHigh,num.DownReg_pCO2History.LowHigh) # sum of DEGs with the criteria pdj < 0.05 + LFC>1 (< -1)
# Write results - covert to as.data.frame for the ordered results
resdata_pCO2History.LowHigh            <- merge(as.data.frame(DEGresults.ordered_pCO2History.LowHigh), 
                                                as.data.frame(counts(dds_2filt, normalized=TRUE)), 
                                                by="row.names", 
                                                sort=FALSE) %>% dplyr::filter(padj < 0.05)## Merge with normalized count data
names(resdata_pCO2History.LowHigh)[1] <- "Airradians_TranscriptID"
resdata.annot_pCO2History.LowHigh     <- merge(Airr_Cvirg_annotation, resdata_pCO2History.LowHigh, 
                                               by = 'Airradians_TranscriptID') 
# View(resdata.annot_pCO2History.LowHigh)
nrow(resdata.annot_pCO2History.LowHigh) == total_pCO2History.LowHigh
resdata.annot_pCO2History.LowHigh     <- resdata.annot_pCO2History.LowHigh[order(
                                             resdata.annot_pCO2History.LowHigh$padj),]    %>% # Order by adjusted
                                                dplyr::filter(!abs(log2FoldChange) < 1) %>% 
                                                dplyr::mutate(Dir = case_when(log2FoldChange > 1 ~ "up",
                                                                              log2FoldChange < 1 ~ "down"),
                                                              About = "Low_v_High")
# write out annotated DESeq2 DEGs
write.csv(resdata.annot_pCO2History.LowHigh, "Output/Transcriptomics/DESeq2/pCO2History_Low_v_High.csv") # write


# ========================================================== 
#  CO2 Cohorts - Low v. Moderate   
# ==========================================================
# Grab DEGs from  the 'dds.run' Moderate
DEGresults_pCO2History.LowModerate         <- results(dds_2filt, 
                                                  contrast=c("pCO2_history", 
                                                             "low", "moderate"),
                                                  alpha = 0.05) # FDR is the alpha 0.05

numDEGs.pval_pCO2History.LowModerate       <- data.frame(table(
                                                      DEGresults_pCO2History.LowModerate$pvalue<0.05)
                                                     )[2,2] # DEGs - NOT considering LFC - just p adj


numDEGs.padj_pCO2History.LowModerate       <- data.frame(table(
                                               DEGresults_pCO2History.LowModerate$padj<0.05))[2,2] 


DEGresults.ordered_pCO2History.LowModerate <- DEGresults_pCO2History.LowModerate[order(
                                                  DEGresults_pCO2History.LowModerate$padj), ] # Order by adjusted p-value

num.UpReg_pCO2History.LowModerate          <- sum((DEGresults.ordered_pCO2History.LowModerate$log2FoldChange
                                               [1:numDEGs.padj_pCO2History.LowModerate] > 0) == TRUE) #  LFC >= 1

num.DownReg_pCO2History.LowModerate        <- sum((DEGresults.ordered_pCO2History.LowModerate$log2FoldChange
                                               [1:numDEGs.padj_pCO2History.LowModerate] < 0) == TRUE) # LFC >= 1

total_pCO2History.LowModerate              <- sum(num.UpReg_pCO2History.LowModerate,num.DownReg_pCO2History.LowModerate) # sum of DEGs with the criteria pdj < 0.05 + LFC>1 (< -1)
# Write results - covert to as.data.frame for the ordered results
resdata_pCO2History.LowModerate            <- merge(as.data.frame(DEGresults.ordered_pCO2History.LowModerate), 
                                                as.data.frame(counts(dds_2filt, normalized=TRUE)), 
                                                by="row.names", 
                                                sort=FALSE) %>% dplyr::filter(padj < 0.05)## Merge with normalized count data
names(resdata_pCO2History.LowModerate)[1] <- "Airradians_TranscriptID"
resdata.annot_pCO2History.LowModerate     <- merge(Airr_Cvirg_annotation,resdata_pCO2History.LowModerate, 
                                               by = 'Airradians_TranscriptID') 
# View(resdata.annot_pCO2History.LowModerate)
nrow(resdata.annot_pCO2History.LowModerate) == total_pCO2History.LowModerate
resdata.annot_pCO2History.LowModerate     <- resdata.annot_pCO2History.LowModerate[order(
                                             resdata.annot_pCO2History.LowModerate$padj),]    %>% # Order by adjusted
                                                dplyr::filter(!abs(log2FoldChange) < 1) %>% 
                                                dplyr::mutate(Dir = case_when(log2FoldChange > 1 ~ "up",
                                                                              log2FoldChange < 1 ~ "down"),
                                                              About = "Low_v_Moderate")
# write out annotated DESeq2 DEGs
write.csv(resdata.annot_pCO2History.LowModerate, "Output/Transcriptomics/DESeq2/pCO2History_Low_v_Moderate.csv") # write



# ========================================================== 
#  CO2 Cohorts - Moderate v. High   
# ==========================================================
# Grab DEGs from  the 'dds.run' High
DEGresults_pCO2History.ModerateHigh         <- results(dds_2filt, 
                                                  contrast=c("pCO2_history", 
                                                             "moderate", "severe"),
                                                  alpha = 0.05) # FDR is the alpha 0.05

numDEGs.pval_pCO2History.ModerateHigh       <- data.frame(table(
                                                      DEGresults_pCO2History.ModerateHigh$pvalue<0.05)
                                                     )[2,2] # DEGs - NOT considering LFC - just p adj


numDEGs.padj_pCO2History.ModerateHigh       <- data.frame(table(
                                               DEGresults_pCO2History.ModerateHigh$padj<0.05))[2,2] 


DEGresults.ordered_pCO2History.ModerateHigh <- DEGresults_pCO2History.ModerateHigh[order(
                                                  DEGresults_pCO2History.ModerateHigh$padj), ] # Order by adjusted p-value

num.UpReg_pCO2History.ModerateHigh          <- sum((DEGresults.ordered_pCO2History.ModerateHigh$log2FoldChange
                                               [1:numDEGs.padj_pCO2History.ModerateHigh] > 0) == TRUE) #  LFC >= 1

num.DownReg_pCO2History.ModerateHigh        <- sum((DEGresults.ordered_pCO2History.ModerateHigh$log2FoldChange
                                               [1:numDEGs.padj_pCO2History.ModerateHigh] < 0) == TRUE) # LFC >= 1

total_pCO2History.ModerateHigh              <- sum(num.UpReg_pCO2History.ModerateHigh,num.DownReg_pCO2History.ModerateHigh) # sum of DEGs with the criteria pdj < 0.05 + LFC>1 (< -1)
# Write results - covert to as.data.frame for the ordered results
resdata_pCO2History.ModerateHigh            <- merge(as.data.frame(DEGresults.ordered_pCO2History.ModerateHigh), 
                                                as.data.frame(counts(dds_2filt, normalized=TRUE)), 
                                                by="row.names", 
                                                sort=FALSE) %>% dplyr::filter(padj < 0.05)## Merge with normalized count data
names(resdata_pCO2History.ModerateHigh)[1] <- "Airradians_TranscriptID"
resdata.annot_pCO2History.ModerateHigh     <- merge(Airr_Cvirg_annotation,resdata_pCO2History.ModerateHigh, 
                                               by = 'Airradians_TranscriptID') 
# View(resdata.annot_pCO2History.ModerateHigh)
nrow(resdata.annot_pCO2History.ModerateHigh) == total_pCO2History.ModerateHigh
resdata.annot_pCO2History.ModerateHigh     <- resdata.annot_pCO2History.ModerateHigh[order(
                                             resdata.annot_pCO2History.ModerateHigh$padj),]   %>% # Order by adjusted
                                                dplyr::filter(!abs(log2FoldChange) < 1) %>% 
                                                dplyr::mutate(Dir = case_when(log2FoldChange > 1 ~ "up",
                                                                              log2FoldChange < 1 ~ "down"),
                                                              About = "Moderate_v_High")

# write out annotated DESeq2 DEGs
write.csv(resdata.annot_pCO2History.ModerateHigh, "Output/Transcriptomics/DESeq2/pCO2History_Moderate_v_High.csv") # write

```




### VENN Diagram
```{r}
library(dplyr)
library(VennDiagram)
library("ggVennDiagram")
library(ggvenn)
library(gridExtra)
library(eulerr)


DEGs_LowHigh      <- resdata.annot_pCO2History.LowHigh %>% 
                                    dplyr::select(Airradians_TranscriptID, 
                                                  log2FoldChange,blastxEval_CvirgGeneID, blastxEval_CvirgProteinID) %>%
                                    dplyr::filter(!abs(log2FoldChange) < 1) %>% 
                                    dplyr::mutate(Dir = case_when(log2FoldChange > 1 ~ "up",
                                                                  log2FoldChange < 1 ~ "down"),
                                                  About = "Low_v_High")

DEGs_LowModerate  <- resdata.annot_pCO2History.LowModerate %>% 
                                    dplyr::select(Airradians_TranscriptID, 
                                                  log2FoldChange,blastxEval_CvirgGeneID, blastxEval_CvirgProteinID) %>%
                                    dplyr::filter(!abs(log2FoldChange) < 1) %>% 
                                    dplyr::mutate(Dir = case_when(log2FoldChange > 1 ~ "up",
                                                                  log2FoldChange < 1 ~ "down"),
                                                  About = "Low_v_Moderate")

DEGs_ModerateHigh <- resdata.annot_pCO2History.ModerateHigh %>% 
                                    dplyr::select(Airradians_TranscriptID, 
                                                  log2FoldChange,blastxEval_CvirgGeneID, blastxEval_CvirgProteinID) %>%
                                    dplyr::filter(!abs(log2FoldChange) < 1) %>% 
                                    dplyr::mutate(Dir = case_when(log2FoldChange > 1 ~ "up",
                                                                  log2FoldChange < 1 ~ "down"),
                                                  About = "Moderate_v_High")

# Merge
bind1       <- rbind(DEGs_LowHigh, DEGs_LowModerate)
Master_DEGs <- rbind(bind1, DEGs_ModerateHigh)




Low_cohort_master <- list(
  vHigh.UP      = (Master_DEGs %>% dplyr::filter(About %in% 'Low_v_High' & 
                                                  Dir %in% 'up'))$Airradians_TranscriptID, 
  vMod.UP      = (Master_DEGs %>% dplyr::filter(About %in% 'Low_v_Moderate' & 
                                                  Dir %in% 'up'))$Airradians_TranscriptID,
  vHigh.DOWN      = (Master_DEGs %>% dplyr::filter(About %in% 'Low_v_High' & 
                                                  Dir %in% 'down'))$Airradians_TranscriptID, 
  vMod.DOWN      = (Master_DEGs %>% dplyr::filter(About %in% 'Low_v_Moderate' & 
                                                  Dir %in% 'down'))$Airradians_TranscriptID
)


Mod_cohort_master <- list(
  vLow.UP      = (Master_DEGs %>% dplyr::filter(About %in% 'Low_v_Moderate' & 
                                                  Dir %in% 'down'))$Airradians_TranscriptID, 
  vHigh.UP     = (Master_DEGs %>% dplyr::filter(About %in% 'Moderate_v_High' & 
                                                  Dir %in% 'up'))$Airradians_TranscriptID,
  vLow.DOWN    = (Master_DEGs %>% dplyr::filter(About %in% 'Low_v_Moderate' & 
                                                 Dir %in% 'up'))$Airradians_TranscriptID, 
  vHigh.DOWN   = (Master_DEGs %>% dplyr::filter(About %in% 'Moderate_v_High' & 
                                                  Dir %in% 'down'))$Airradians_TranscriptID
)


High_cohort_master <- list(
  vLow.UP      = (Master_DEGs %>% dplyr::filter(About %in% 'Low_v_High' & 
                                                  Dir %in% 'down'))$Airradians_TranscriptID, 
  vMod.UP      = (Master_DEGs %>% dplyr::filter(About %in% 'Moderate_v_High' & 
                                                  Dir %in% 'down'))$Airradians_TranscriptID,
  vLow.DOWN      = (Master_DEGs %>% dplyr::filter(About %in% 'Low_v_High' & 
                                                  Dir %in% 'up'))$Airradians_TranscriptID, 
  vMod.DOWN      = (Master_DEGs %>% dplyr::filter(About %in% 'Moderate_v_High' & 
                                                  Dir %in% 'up'))$Airradians_TranscriptID
)


All <- list(
  `H>M`      = unique((Master_DEGs %>% dplyr::filter(About %in% 'Moderate_v_High' & 
                                                  Dir %in% 'down'))$Airradians_TranscriptID),
  `H>L`      = unique((Master_DEGs %>% dplyr::filter(About %in% 'Low_v_High' & 
                                                  Dir %in% 'down'))$Airradians_TranscriptID), 
  `M>L`      = unique((Master_DEGs %>% dplyr::filter(About %in% 'Low_v_Moderate' & 
                                                  Dir %in% 'down'))$Airradians_TranscriptID),
  `M>H`      = unique((Master_DEGs %>% dplyr::filter(About %in% 'Moderate_v_High' & 
                                                  Dir %in% 'up'))$Airradians_TranscriptID),
  `L>H`      = unique((Master_DEGs %>% dplyr::filter(About %in% 'Low_v_High' & 
                                                  Dir %in% 'up'))$Airradians_TranscriptID), 
  `L>M`      = unique((Master_DEGs %>% dplyr::filter(About %in% 'Low_v_Moderate' & 
                                                  Dir %in% 'up'))$Airradians_TranscriptID)
)

library(ggVennDiagram)
All_Venn              <- ggVennDiagram(All, 
                                        color = 1, 
                                        lwd = 0.7, 
                                        label = "count")  + 
                                  scale_fill_gradient(low = "white", high = "#4981BF")  + 
                                  ggtitle("Low pCO2 Cohort response to Elevated pCO2") +
                                  theme(legend.position = "none")


All_Venn_quantitites  <- plot(euler(All, shape = "ellipse"), quantities = TRUE)


pdf("Output/Transcriptomics/DESeq2/Cohorts_DEGs_Venn.pdf", width=6, height=12)
ggpubr::ggarrange(All_Venn,
          All_Venn_quantitites,
          nrow=2)
dev.off()


# 64 genes that low downregulated relative to moderate and high (necessary under elevated pCO2?)
# motor proteins (unconventional myosin-XV, dynein gamma chain, flagellar outer arm), ubiquitination oproteasomal degredation
# 
LowpCO2_cohort_Down <- Master_DEGs %>%
                           dplyr::filter(About %in% c('Low_v_High','Low_v_Moderate') &
                                         Dir %in% 'down') %>% 
                       dplyr::mutate(log2FoldChange = abs(log2FoldChange)) %>% 
                       dplyr::arrange(desc(log2FoldChange)) %>% 
                       dplyr::filter(!blastxEval_CvirgGeneID %in% 'uncharacterized') %>% 
                       # na.omit() %>% 
                       group_by(Airradians_TranscriptID) %>%
                       add_count(name = "id_occurrence") %>% 
                       dplyr::filter(id_occurrence == 2) %>% 
                       dplyr::select(c(Airradians_TranscriptID, blastxEval_CvirgGeneID, blastxEval_CvirgProteinID, log2FoldChange))

LowpCO2_cohort_Down <- setDT(LowpCO2_cohort_Down)[, list(log2FoldChange=if(.N==2) median(log2FoldChange) else min(log2FoldChange)) , 
                           by = LowpCO2_cohort_Down[,1:3]] %>%
                           dplyr::rename(MEDIANlog2FoldChange = log2FoldChange) %>% 
                           dplyr::arrange(desc(MEDIANlog2FoldChange))

nrow(LowpCO2_cohort_Down)

write.csv(LowpCO2_cohort_Down, "Output/Transcriptomics/DESeq2/Venn_intersections/Low_expression_LowpCO2.csv")
# 6 genes that Moderate expressed higher than low and high 
# - ceruloplasmin-like, far upstream element-binding protein 1-like, multidrug resistance protein 1-like
ModpCO2_cohort_Up_1 <- Master_DEGs %>% dplyr::filter(About %in% 'Low_v_Moderate' & Dir %in% 'down')
ModpCO2_cohort_Up_2 <- Master_DEGs %>% dplyr::filter(About %in% 'Moderate_v_High' & Dir %in% 'up')
ModpCO2_cohort_Up   <- rbind(ModpCO2_cohort_Up_1, ModpCO2_cohort_Up_2) %>% 
                       dplyr::mutate(log2FoldChange = abs(log2FoldChange)) %>% 
                       dplyr::arrange(desc(log2FoldChange)) %>% 
                       dplyr::filter(!blastxEval_CvirgGeneID %in% 'uncharacterized') %>% 
                       na.omit() %>% 
                       group_by(Airradians_TranscriptID) %>%
                       add_count(name = "id_occurrence") %>% 
                       dplyr::filter(id_occurrence == 2) %>% 
                       dplyr::select(c(Airradians_TranscriptID, blastxEval_CvirgGeneID, blastxEval_CvirgProteinID, log2FoldChange))

ModpCO2_cohort_Up <- setDT(ModpCO2_cohort_Up)[, list(log2FoldChange=if(.N==2) median(log2FoldChange) else min(log2FoldChange)) , 
                           by = ModpCO2_cohort_Up[,1:3]] %>%
                           dplyr::rename(MEDIANlog2FoldChange = log2FoldChange) %>% 
                           dplyr::arrange(desc(MEDIANlog2FoldChange))

nrow(ModpCO2_cohort_Up)
  
write.csv(ModpCO2_cohort_Up, "Output/Transcriptomics/DESeq2/Venn_intersections/High_expression_ModeratepCO2.csv")

# 4 genes that High had downregulated relative to low and moderate 
# - cytochrome P450 2C8-like
HighpCO2_cohort_DOWN <- Master_DEGs %>% dplyr::filter(About %in% c('Low_v_High','Moderate_v_High') & 
                                                          Dir %in% 'up') %>% 
                       dplyr::mutate(log2FoldChange = abs(log2FoldChange)) %>% 
                       dplyr::arrange(desc(log2FoldChange)) %>% 
                       dplyr::filter(!blastxEval_CvirgGeneID %in% 'uncharacterized') %>% 
                       na.omit() %>% 
                       group_by(Airradians_TranscriptID) %>%
                       add_count(name = "id_occurrence") %>% 
                       dplyr::filter(id_occurrence == 2) %>% 
                       dplyr::select(c(Airradians_TranscriptID, blastxEval_CvirgGeneID, blastxEval_CvirgProteinID, log2FoldChange))

HighpCO2_cohort_DOWN <- setDT(HighpCO2_cohort_DOWN)[, list(log2FoldChange=if(.N==2) median(log2FoldChange) else min(log2FoldChange)) , 
                           by = HighpCO2_cohort_DOWN[,1:3]] %>%
                           dplyr::rename(MEDIANlog2FoldChange = log2FoldChange) %>% 
                           dplyr::arrange(desc(MEDIANlog2FoldChange))
nrow(HighpCO2_cohort_DOWN)

write.csv(HighpCO2_cohort_DOWN, "Output/Transcriptomics/DESeq2/Venn_intersections/Low_expression_HighpCO2.csv")
# 5 genes that moderate downregulated relative to to low and high
# - WD repeat-containing protein 6-like, biogenesis of lysosome-related organelles complex 1 subunit 1-like
ModpCO2_cohort_Down_1 <- Master_DEGs %>% dplyr::filter(About %in% 'Low_v_Moderate' & Dir %in% 'up')
ModpCO2_cohort_Down_2 <- Master_DEGs %>% dplyr::filter(About %in% 'Moderate_v_High' & Dir %in% 'down')
ModpCO2_cohort_Down   <- rbind(ModpCO2_cohort_Down_1, ModpCO2_cohort_Down_2) %>% 
                       dplyr::mutate(log2FoldChange = abs(log2FoldChange)) %>% 
                       dplyr::arrange(desc(log2FoldChange)) %>% 
                       dplyr::filter(!blastxEval_CvirgGeneID %in% 'uncharacterized') %>% 
                       na.omit() %>% 
                       group_by(Airradians_TranscriptID) %>%
                       add_count(name = "id_occurrence") %>% 
                       dplyr::filter(id_occurrence == 2) %>% 
                       dplyr::select(c(Airradians_TranscriptID, blastxEval_CvirgGeneID, blastxEval_CvirgProteinID, log2FoldChange))
  
ModpCO2_cohort_Down <- setDT(ModpCO2_cohort_Down)[, list(log2FoldChange=if(.N==2) median(log2FoldChange) else min(log2FoldChange)) , 
                           by = ModpCO2_cohort_Down[,1:3]] %>%
                           dplyr::rename(MEDIANlog2FoldChange = log2FoldChange) %>% 
                           dplyr::arrange(desc(MEDIANlog2FoldChange))

nrow(ModpCO2_cohort_Down)
write.csv(ModpCO2_cohort_Down, "Output/Transcriptomics/DESeq2/Venn_intersections/Low_expression_ModeratepCO2.csv")


# 4 genes that High had downregulated relative to low and moderate 
# - cytochrome P450 2C8-like
HighpCO2_cohort_UP <- Master_DEGs %>% dplyr::filter(About %in% c('Low_v_High','Moderate_v_High') & 
                                                          Dir %in% 'down') %>% 
                       dplyr::mutate(log2FoldChange = abs(log2FoldChange)) %>% 
                       dplyr::arrange(desc(log2FoldChange)) %>% 
                       dplyr::filter(!blastxEval_CvirgGeneID %in% 'uncharacterized') %>% 
                       na.omit() %>% 
                       group_by(Airradians_TranscriptID) %>%
                       add_count(name = "id_occurrence") %>% 
                       dplyr::filter(id_occurrence == 2) %>% 
                       dplyr::select(c(Airradians_TranscriptID, blastxEval_CvirgGeneID, blastxEval_CvirgProteinID, log2FoldChange))

HighpCO2_cohort_UP <- setDT(HighpCO2_cohort_UP)[, list(log2FoldChange=if(.N==2) median(log2FoldChange) else min(log2FoldChange)), 
                           by = HighpCO2_cohort_UP[,1:3]] %>%
                           dplyr::rename(MEDIANlog2FoldChange = log2FoldChange) %>% 
                           dplyr::arrange(desc(MEDIANlog2FoldChange))
nrow(HighpCO2_cohort_UP)

write.csv(HighpCO2_cohort_UP, "Output/Transcriptomics/DESeq2/Venn_intersections/High_expression_HighpCO2.csv")
```
