---
title: "DESeq2"
author: "Samuel Gurr"
date: "2024-02-21"
output: pdf_document
---


### Set working directory
```{r setup, include=TRUE}
knitr::opts_chunk$set(echo = TRUE, 
                      warning = FALSE, 
                      message = FALSE, 
                      cache = TRUE)

knitr::opts_knit$set(root.dir = "C:/Users/samjg/Documents/Github_repositories/Airradians_CellularMolecular_OA/RAnalysis")
```

### Load libraries
```{r load_libraries, include = TRUE}
library(dplyr)
library(reshape2)
library(ggplot2)
library(kableExtra)
library(knitr)
library(tidyverse)
library(DESeq2)
library(plotrix)
library(ggpubr)
```


### Load data: choose ONE of the counts clusters (filtered raw reads or unfiltered raw reads)

### Sample metadata - Experimental treatments/groups
```{r experiment_data, include = TRUE}
# load data and read in the module membership master file 
load(file = "Output/Transcriptomics/WGCNA/challenge/Challenge-networkConstruction-stepByStep.RData")


Challenge_ModuleMembership  <- read.csv(file="Output/Transcriptomics/WGCNA/challenge/Challenge_WGCNA_ModulMembership.csv", 
                                        sep=',', header=TRUE)

AirrdiansTranscript_Proteinnames <- Challenge_ModuleMembership %>% select(Airradians.TranscriptID,Protein_name)
# transpose the vst challenge data used for WGCNa (jas 27 samples - truncated from sample tree (view WGCNA.Rmd)
vstExp_Challenge  <- as.data.frame(t(dds_1filt_master)) # this is our vst master file we used for WGCNA
vstExp_Challenge  <- tibble::rownames_to_column(vstExp_Challenge,"Airradians.TranscriptID") # rownames as first column
ncol(vstExp_Challenge) # 27 + AirradiansID
# write out this rlog tranformed master data (used for plotting below1) 
rlogExp_Challenge <- read.csv(file = "Output/Transcriptomics/WGCNA/challenge/Challenge_rlog_transformed.csv", sep=',', header=TRUE) %>% 
                          dplyr::select(!X)

ncol(rlogExp_Challenge) # 27 + AirradiansID

```

```{r call the count matrices, include = TRUE}

moduleRed_Airradians.IDs            <- (Challenge_ModuleMembership %>% dplyr::filter(moduleColor %in%
                                                                                       'red'))$Airradians.TranscriptID
#vst 
vstExp_Challenge.long              <- vstExp_Challenge %>% 
                                            dplyr::filter(Airradians.TranscriptID %in% moduleRed_Airradians.IDs) %>% 
                                            pivot_longer(
                                              cols = !c(Airradians.TranscriptID), 
                                              names_to = "Sample_num",
                                              values_to = "vstExp"
                                            )
nrow(vstExp_Challenge.long) # 15066

vstExp_Challenge.long.merge <- merge(vstExp_Challenge.long, AirrdiansTranscript_Proteinnames, by = 'Airradians.TranscriptID')

vstExp_Challenge.long.meta         <- cbind(vstExp_Challenge.long.merge, exp_metadata_1_master, by = "Sample_num")

vstExp_Challenge.long.meta.means   <- vstExp_Challenge.long.meta %>% 
                                            dplyr::select(c(vstExp, pCO2_exposure,pCO2_history, 
                                                            Airradians.TranscriptID, Protein_name)) %>% 
                                            dplyr::group_by(pCO2_exposure, pCO2_history, Airradians.TranscriptID,Protein_name) %>% 
                                            dplyr::summarise(mean.vstExp = mean(vstExp), n = n()) 

nrow(vstExp_Challenge.long.meta.means)


#rlog 
rlogExp_Challenge.long              <- rlogExp_Challenge %>% 
                                            dplyr::filter(Airradians.TranscriptID %in% moduleRed_Airradians.IDs) %>% 
                                            pivot_longer(
                                              cols = !c(Airradians.TranscriptID), 
                                              names_to = "Sample_num",
                                              values_to = "rlogExp"
                                            )
rlogExp_Challenge.long.meta         <- cbind(rlogExp_Challenge.long, exp_metadata_1_master, by = "Sample_num")

rlogExp_Challenge.long.meta.means   <- rlogExp_Challenge.long.meta %>% 
                                            dplyr::select(c(rlogExp, pCO2_exposure, 
                                                            pCO2_history, Airradians.TranscriptID)) %>% 
                                            dplyr::group_by(pCO2_exposure, pCO2_history, Airradians.TranscriptID) %>% 
                                            dplyr::summarise(mean.rlogExp = mean(rlogExp), n = n()) 
# View(rlogExp_LowvHigh.long.meta.means)

```


```{r, loop to assign frontloaded genes (RESPONSE TO HIGH)}
loop_df    <- as.data.frame(unique(vstExp_Challenge.long.meta.means$Airradians.TranscriptID))
# loop_df    <- as.data.frame(unique(rlogExp_Challenge.long.meta.means$Airradians.TranscriptID))
loop.table <- data.frame(matrix(nrow = 1, ncol = 9)) # create dataframe to save cumunalitively during for loop
colnames(loop.table)<-c('Airradians.TranscriptID',
                        'Protein_name',
                        'baseMeanNAIVE_control', 
                        'baseMeanHABITUATED_control', 
                        'baseMeanNAIVE_response', 
                        'baseMeanHABITUATED_response', 
                        'ControlRatio',
                        'ResponseRatioHigh',
                        'Frontloaded_criteria')

df_total <- data.frame() # start dataframe 
for (i in 1:nrow(loop_df)) {
  
  df_loop <- vstExp_Challenge.long.meta.means %>% filter(Airradians.TranscriptID %in% loop_df[i,])
  # df_loop <- rlogExp_Challenge.long.meta.means %>% filter(Airradians.TranscriptID %in% loop_df[i,])
  if (nrow(df_loop) == 6) {
  
  loop.table$Airradians.TranscriptID     <- loop_df[i,]
  loop.table$Protein_name                <- df_loop$Protein_name[1]
  # loop.table$baseMeanNAIVE_control       <- (df_loop %>% dplyr::filter(pCO2_exposure %in% 'low' & pCO2_history %in% 'low'))$mean.rlogExp
  # loop.table$baseMeanHABITUATED_control  <- (df_loop %>% dplyr::filter(pCO2_exposure %in% 'low' & pCO2_history %in% 'severe'))$mean.rlogExp
  # 
  # loop.table$baseMeanNAIVE_response      <- (df_loop %>% dplyr::filter(pCO2_exposure %in% 'severe' & pCO2_history %in% 'low'))$mean.rlogExp
  # loop.table$baseMeanHABITUATED_response <- (df_loop %>% dplyr::filter(pCO2_exposure %in% 'severe' & pCO2_history %in% 'severe'))$mean.rlogExp
  
  
  loop.table$baseMeanNAIVE_control       <- (df_loop %>% dplyr::filter(pCO2_exposure %in% 'low' & 
                                                                         pCO2_history %in% 'low'))$mean.vstExp
  loop.table$baseMeanHABITUATED_control  <- (df_loop %>% dplyr::filter(pCO2_exposure %in% 'low' & 
                                                                         pCO2_history %in% 'severe'))$mean.vstExp
  
  loop.table$baseMeanNAIVE_response      <- (df_loop %>% dplyr::filter(pCO2_exposure %in% 'severe' & 
                                                                         pCO2_history %in% 'low'))$mean.vstExp
  loop.table$baseMeanHABITUATED_response <- (df_loop %>% dplyr::filter(pCO2_exposure %in% 'severe' & 
                                                                         pCO2_history %in% 'severe'))$mean.vstExp
  
  loop.table$ControlRatio                <- loop.table$baseMeanHABITUATED_control / loop.table$baseMeanNAIVE_control
  
  loop.table$ResponseRatioHigh           <- ( loop.table$baseMeanHABITUATED_response / 
                                                loop.table$baseMeanHABITUATED_control ) /
    
                                            ( loop.table$baseMeanNAIVE_response / 
                                                loop.table$baseMeanNAIVE_control)
  
  loop.table <- loop.table %>% 
                          dplyr::mutate(Frontloaded_criteria = 
                                          case_when(
                                            ControlRatio > 1 & 
                                              ResponseRatioHigh < 1 ~ "frontloaded",
                                            
                                            ControlRatio < 1.00 & 
                                              ResponseRatioHigh < 1.00 ~ "low expression relative to naive",
                                            
                                            ControlRatio < 1.00 & 
                                              ResponseRatioHigh > 1.00 ~ "more responsive relative to naive"))
  } else {NA}
  df <- data.frame(loop.table) # name dataframe for this singl e row
  df_total <- rbind(df_total,df) #bind to a cumulative list dataframe
}    
View( (df_total %>% filter(Frontloaded_criteria == 'frontloaded')))

WGCNA_frontloaded <- (df_total %>% filter(Frontloaded_criteria == 'frontloaded'))

View(WGCNA_frontloaded %>% filter(Airradians.TranscriptID %in% DESEq2_frontloaded$Airradians_TranscriptID))

Frontloading_smooth <- df_total %>% 
                        dplyr::mutate(FrontMod_color = 
                                        ifelse(ResponseRatioHigh < 1.0 & 
                                                 ControlRatio > 1.0, "True", "False")) %>%
                        # dplyr::filter(ControlRatio < 5) %>% 
                        # dplyr::filter(ResponseRatioHigh <1.2) %>% 
                                      ggplot(aes(x=ResponseRatioHigh, y=ControlRatio)) +
                                      geom_point(aes(color='grey80', na.rm=TRUE)) +  
                                         scale_shape_manual(values=c(19,19)) + 
                                         scale_color_manual(values=c("grey", "black")) +
                                      theme_classic() + 
                                      scale_x_continuous(expand = c(0, 0), limits = c(0.9,1.1)) +
                                      scale_y_continuous(expand = c(0, 0), limits = c(0.95,1.1)) +
                                      stat_smooth(method = "lm", 
                                                  formula = y ~ x + poly(x, 2) - 1) +
                                      geom_vline(xintercept=1, linetype="dotted") + 
                                      geom_hline(yintercept=1, linetype="dotted") + 
                                      labs(y= "Low pCO2 exposure ratio", 
                                           x = "Response ratio to high pCO2",
                                           title = "Frontloading criteria (shaded)") + 
                                      expand_limits(x = 1.1, y = 1.1) + 
                                      annotate("rect", xmin = 0.9, xmax = 1, ymin = 1, ymax = 1.1,
                                               alpha = .2) + 
                                      theme(legend.position="none", text = element_text(size=10))

# truncate df_total for only genes assigned as 'frontloaded' from module red
Frontloaded_AirradiansTranscriptIDs <- df_total %>% filter(Frontloaded_criteria %in% 'frontloaded')

write.csv(Frontloaded_AirradiansTranscriptIDs, file = "Output/Transcriptomics/WGCNA/challenge/WGCNA_Frontloaded_transcripts_modulered.csv")



# truncate the vst Exp means (mean exp for each gene within pCO2 history * pCO2 exposure, N = 5 each)
IDtargets                           <- vstExp_Challenge.long.meta.means %>% 
                                          dplyr::filter(Airradians.TranscriptID %in%
                                          unique(Frontloaded_AirradiansTranscriptIDs$Airradians.TranscriptID))

IDtargets_stats <- IDtargets %>% dplyr::filter(pCO2_exposure %in% c('low', 'severe'))
mod <- lm(mean.vstExp~pCO2_history*pCO2_exposure,data=IDtargets_stats)
shapiro.test(resid(mod)) # p-value < 2.2e-16
summary(aov(mod))
library(Rmisc)
library(rcompanion)
mod <- scheirerRayHare(mean.vstExp~pCO2_history*pCO2_exposure,data=IDtargets_stats)
mod

plotting <- IDtargets %>% group_by(pCO2_history, pCO2_exposure) %>%   
                          dplyr::summarise(mean_meanvstExp = mean(mean.vstExp),
                            sd_vstExp = sd(mean.vstExp),
                            n = n(),
                            se_vstExp = sd_vstExp / sqrt(n)
                          )
Frontloading_MEAN.SE <- plotting %>% 
        dplyr::filter(pCO2_exposure %in% c('low', 'severe')) %>% 
             ggplot(aes(x=pCO2_exposure, y = mean_meanvstExp, fill=pCO2_history)) +
                     geom_point(aes(shape=pCO2_history, fill=pCO2_history), 
                                    size = 4.5,position=position_dodge(.4)) + 
               geom_line() +
               theme_classic() +
               labs(y= "vst expression", 
                    x = "pCO2 exposure",
                    title = "Frontloaded genes") + 
               geom_errorbar(aes(ymin=(mean_meanvstExp)-(se_vstExp), # new means and se by treatment
                                 ymax=(mean_meanvstExp)+(se_vstExp)), # new means and se by treatment
                                 width=0,position=position_dodge(.4)) # width determines the length of the end ticks

pdf("Output/Transcriptomics/WGCNA/challenge/WGCNA_Frontloaded_transcripts_modulered.pdf", height=4, width =8)
ggarrange(Frontloading_smooth, Frontloading_MEAN.SE, ncol=2)
dev.off()
```




```{r, loop to assign frontloaded genes RESPONSE TO MODERATE}
loop_df    <- as.data.frame(unique(rlogExp_LowvHigh.long.meta.means$Airradians_TranscriptID))
loop.table <- data.frame(matrix(nrow = 1, ncol = 8)) # create dataframe to save cumunalitively during for loop
colnames(loop.table)<-c('Airradians_TranscriptID',
                        'baseMeanNAIVE_control', 
                        'baseMeanHABITUATED_control', 
                        'baseMeanNAIVE_response', 
                        'baseMeanHABITUATED_response', 
                        'ControlRatio',
                        'ResponseRatioModerate',
                        'Frontloaded_criteria')

df_total <- data.frame() # start dataframe 
for (i in 1:nrow(loop_df)) {
  
  df_loop <- rlogExp_LowvHigh.long.meta.means %>% filter(Airradians_TranscriptID %in% loop_df[i,])
  
  if (nrow(df_loop) == 6) {
  
  loop.table$Airradians_TranscriptID     <- loop_df[i,]
  
  loop.table$baseMeanNAIVE_control       <- (df_loop %>% dplyr::filter(pCO2_exposure %in% 'low' & 
                                                                         pCO2_history %in% 'low'))$mean.rlogExp
  
  loop.table$baseMeanHABITUATED_control  <- (df_loop %>% dplyr::filter(pCO2_exposure %in% 'low' & 
                                                                         pCO2_history %in% 'severe'))$mean.rlogExp
  
  loop.table$baseMeanNAIVE_response      <- (df_loop %>% dplyr::filter(pCO2_exposure %in% 'moderate' & 
                                                                         pCO2_history %in% 'low'))$mean.rlogExp
  
  loop.table$baseMeanHABITUATED_response <- (df_loop %>% dplyr::filter(pCO2_exposure %in% 'moderate' & 
                                                                         pCO2_history %in% 'severe'))$mean.rlogExp
  
  loop.table$ControlRatio                <- loop.table$baseMeanHABITUATED_control / loop.table$baseMeanNAIVE_control
  loop.table$ResponseRatioModerate       <- ( loop.table$baseMeanHABITUATED_response / loop.table$baseMeanHABITUATED_control ) /
                                            ( loop.table$baseMeanNAIVE_response / loop.table$baseMeanNAIVE_control)
  
  loop.table <- loop.table %>% 
                          dplyr::mutate(Frontloaded_criteria = case_when(ControlRatio > 1.00 & 
                                                                           ResponseRatioModerate< 1.00 ~ "frontloaded", 
                                        ControlRatio < 1.00 & ResponseRatioModerate < 1.00 ~ "low expression relative to naive",
                                        ControlRatio < 1.00 & ResponseRatioModerate > 1.00 ~ "more responsive relative to naive"))
  } else {NA}
  df <- data.frame(loop.table) # name dataframe for this singl e row
  df_total <- rbind(df_total,df) #bind to a cumulative list dataframe
}    
View(df_total)



df_total %>% 
  dplyr::mutate(FrontMod_color = 
                  ifelse(ResponseRatioModerate < 1.0 & 
                           ControlRatio > 1.0, "True", "False")) %>%
  # dplyr::filter(ControlRatio < 5) %>% 
  dplyr::filter(ResponseRatioModerate <1.2) %>% 
                ggplot(aes(x=ResponseRatioModerate, y=ControlRatio)) +
                geom_point(aes(color='grey80', na.rm=TRUE)) +  
                   scale_shape_manual(values=c(19,19)) + 
                   scale_color_manual(values=c("grey", "black")) +
                theme_classic() + 
                scale_x_continuous(expand = c(0, 0), breaks=seq(0,2, 0.2)) +
                scale_y_continuous(expand = c(0, 0), limits = c(0.9,1.4)) +
                stat_smooth(method = "lm", 
                            formula = y ~ x + poly(x, 2) - 1) +
                geom_vline(xintercept=1, linetype="dotted") + 
                geom_hline(yintercept=1, linetype="dotted") + 
                labs(y= "Low pCO2 exposure ratio", 
                     x = "Response ratio to moderate pCO2",
                     title = "Frontloaded genes; response to Moderate pCO2") + 
                expand_limits(x = 1.2, y = 1.4) + 
                annotate("rect", xmin = 0.8, xmax = 1, ymin = 1, ymax = 1.4,
                         alpha = .2) + 
                theme(legend.position="none", text = element_text(size=10))

Frontloaded_AirradiansTranscriptIDs <- df_total %>% filter(Frontloaded_criteria %in% 'frontloaded')
IDtargets <- rlogExp_LowvHigh.long.meta.means %>% dplyr::filter(Airradians_TranscriptID %in% Frontloaded_AirradiansTranscriptIDs$Airradians_TranscriptID)

plotting <- IDtargets %>% group_by(pCO2_history, pCO2_exposure) %>%   
                          summarise(mean_meanrlogExp = mean(mean.rlogExp),
                            sd_rlogExp = sd(mean.rlogExp),
                            n = n(),
                            se_rlogExp = sd_rlogExp / sqrt(n)
                          )

IDtargets_stats <- IDtargets %>% dplyr::filter(pCO2_exposure %in% c('low', 'moderate'))
mod <- lm(mean.rlogExp~pCO2_history*pCO2_exposure,data=IDtargets_stats)
shapiro.test(resid(mod)) # p-value < 2.2e-16
summary(aov(mod))


plot <- plotting %>% dplyr::filter(pCO2_exposure %in% c('low', 'moderate')) %>% 
                                     ggplot(aes(x=pCO2_exposure, y = mean_meanrlogExp, fill=pCO2_history)) +
                                            geom_point(aes(shape=pCO2_history, fill=pCO2_history), size = 4.5,position=position_dodge(.4)) + 
                                            geom_line() +
                                            geom_errorbar(aes(ymin=(mean_meanrlogExp)-(se_rlogExp), # new means and se by treatment
                                                              ymax=(mean_meanrlogExp)+(se_rlogExp)), # new means and se by treatment
                                                            width=0,position=position_dodge(.4)) # width determines the length of the end ticks
plot
```
