---
title: "Frontloading GO"
author: "Samuel Gurr"
date: "May 1, 2022"
output: html_document
---

```{r setup, include=FALSE}
# knitr::opts_knit$set(root.dir = "C:/Users/samjg/Documents/Github_repositories/Airradians_CellularMolecular_OA/RAnalysis")
knitr::opts_knit$set(root.dir = "C:/Users/samuel.gurr/Documents/Github_repositories/Airradians_CellularMolecular_OA/RAnalysis")
```


NOTE: after runinng this once... you can load the libraries (below) and skip to 'the 'Find GOslim terms' to load the .csv files for plotting
Load libraries
```{r load_libraries, include = TRUE}
# load libraries - notes show the install command needed to install (pre installed)
library(goseq)
library(dplyr)
library(forcats)
library(ggplot2)
library(gridExtra)
library(tidyr)
library(grDevices)
library(reshape2)
library(Rmisc)
library(ggpubr)
library(tibble)
library(hrbrthemes)
library(gridExtra)
library(tidyr)
library(zoo)
library(circlize)
library(GSEABase)
library(data.table)
library(stringr)
library(tidyverse)
library(readxl)
```

SETWD & LOAD DATA:
Annotation and DE data (from DESeq2 Rmd script)
```{r Load_annot_and_DE_files, include = TRUE}
# master reference view te master ref R script for details
Master_ref  <- read.csv(file= "Data/Transcriptomics/metadata/Seq_Cvirg_Reference.csv", sep=',', header=TRUE) %>% 
                     dplyr::select(c('Cvirginica_TranscriptID','Annotation_GO_ID', 'Cvirginica_length'))


Cvirginica_annot_reference  <- read.csv(file="Data/Transcriptomics/metadata/seq_id_AirrCvirg_MERGED_master.csv", sep=',', header=TRUE) %>% 
                                  # dplyr::select(c('TranscriptID','Function','GeneID')) %>% 
                                  dplyr::select(c('Airradians_TranscriptID','blastxEval_CvirgProteinID', 
                                                  'blastxEval_CvirgGOterms', 'meanLength')) %>% 
                                  dplyr::mutate(TranscriptID = gsub(" ", "", Airradians_TranscriptID))
                                  # dplyr::mutate(Protein_name = gsub("\\s\\(LOC.*|\\sLOC111.*", "", perl=TRUE, Function)) %>% 
                                  # dplyr::select(!Function)


# load the frotnaoded gene set
Frontloaded_UP    <- read.csv("Output/Transcriptomics/WGCNA/challenge/Frontloading/UPmodulered_Severe.csv")
Frontloaded_DOWN  <- read.csv("Output/Transcriptomics/WGCNA/challenge/Frontloading/DOWNmoduleGreen_Moderate.csv")
# GOslim
slim              <- getOBOCollection("http://current.geneontology.org/ontology/subsets/goslim_generic.obo") #get GO database - # call goslim_generic.obo terms as 'slim'

```


goseq
Prepare dataframe(s) and vectors for goseq 
(1) Format 'GO.term' for goseq from the P.generosa annotation .fna file 'Geoduck_annotation'
```{r GO.terms_data}
# call the GO terms
Cvirginica_GOterms                <- as.data.frame(Cvirginica_annot_reference) %>%
                                      dplyr::select(c('Airradians_TranscriptID','blastxEval_CvirgGOterms'))
colnames(Cvirginica_GOterms)[1:2] <- c('transcript.ID', 'GO.terms') # call gene name and the GO terms - (Uniprot ID 'V5')
splitted                          <- strsplit(as.character(Cvirginica_GOterms$GO.terms), ";") #slit into multiple GO ids by delimiter'; ' remember the space after ; is needed here! without this you will only call the first listed GO term for each gene!
GO.terms                          <- data.frame(v1 = rep.int(Cvirginica_GOterms$transcript.ID, sapply(splitted, length)), 
                                                   v2 = unlist(splitted)) #list all genes with each of their GO terms in a single row
```

```{r DOWN GO.terms_data}
# call the GO terms
Cvirginica_GOterms_DOWN                <- as.data.frame(Frontloaded_DOWN) %>% dplyr::select(c('Airradians.TranscriptID','Annotation_GO_ID'))
colnames(Cvirginica_GOterms_DOWN)[1:2] <- c('transcript.ID', 'GO.terms') # call gene name and the GO terms - (Uniprot ID 'V5')
splitted_DOWN                          <- strsplit(as.character(Cvirginica_GOterms_DOWN$GO.terms), ";") #slit into multiple GO ids by delimiter'; ' remember the space after ; is needed here! without this you will only call the first listed GO term for each gene!
GO.terms_DOWN                          <- data.frame(v1 = rep.int(Cvirginica_GOterms_DOWN$transcript.ID, sapply(splitted_DOWN, length)), 
                                                     v2 = unlist(splitted_DOWN)) #list all genes with each of their GO terms in a single row
```
                                                                                                                                            
(2) Unique Genes - vector based on all unique mapped reads  - Remember this is based on the genes for EACH sampling day becasue DEeq2 was run separately (i.e. days 0, 7, 14, and 21)
(3) Gene length 
Set ID and gene length vectors, and make a binary matrix indicating which genes are differentially expressed. These are used as input to nullp, which for calculates a Probability Weighting Function for each set of DEGs.
```{r UP length vectors}
# Prepare dataframe(s) and vectors for goseq 
# Format 'GO.term' for goseq from the P.generosa annotation .fna file 'Geoduck_annotation'
GO_unique.genes_UP  <- as.vector(unique(Frontloaded_UP$Airradians.TranscriptID)) # call all unique genes for GO analysis (goseq)

# Gene length 
GO_gene.length_UP   <- unique(Frontloaded_UP %>% dplyr::select(c("Airradians.TranscriptID","Cvirginica_length")))

# merge length with counts data
length_vector_UP    <- GO_gene.length_UP$Cvirginica_length

GeneLength.UP       <-  dplyr::inner_join(Frontloaded_UP, 
                                              unique((GO_gene.length_UP %>% 
                                                        dplyr::rename(TranscriptID = Airradians.TranscriptID)), 
                                                     by='TranscriptID'))
# call length values for goseq - confirms that the IDvector and length_vector are the same!!!
sum(sapply(length_vector_UP,length)) == dim(Frontloaded_UP)[1] #should be TRUE

```


```{r DOWN length vectors}
# Prepare dataframe(s) and vectors for goseq 
# Format 'GO.term' for goseq from the P.generosa annotation .fna file 'Geoduck_annotation'
GO_unique.genes_DOWN  <- as.vector(unique(Frontloaded_DOWN$Airradians.TranscriptID)) # call all unique genes for GO analysis (goseq)

# Gene length 
GO_gene.length_DOWN   <- unique(Frontloaded_DOWN %>% dplyr::select(c("Airradians.TranscriptID","Cvirginica_length")))

# merge length with counts data
length_vector_DOWN    <- GO_gene.length_DOWN$Cvirginica_length

GeneLength.DOWN       <-  dplyr::inner_join(Frontloaded_DOWN, 
                                              unique((GO_gene.length_DOWN %>% 
                                                        dplyr::rename(TranscriptID = Airradians.TranscriptID)), 
                                                     by='TranscriptID'))
# call length values for goseq - confirms that the IDvector and length_vector are the same!!!
sum(sapply(length_vector_DOWN,length)) == dim(Frontloaded_DOWN)[1] #should be TRUE

```

It's goseq time!!!
Calculate Probability Weighting Function
```{r UP -  RUN GOSEQ}
GO_unique.genes.all  <- as.vector(unique(Cvirginica_annot_reference$Airradians_TranscriptID)) # call all unique genes for GO analysis (goseq)
GO_gene.length       <- unique(Cvirginica_annot_reference %>% dplyr::select(c("Airradians_TranscriptID","meanLength")))
length_vector        <- GO_gene.length$meanLength



UP_integer <- as.integer(GO_unique.genes.all %in% (Frontloaded_UP$Airradians.TranscriptID)) # w/o day-specific ID vector
names(UP_integer)=GO_unique.genes.all # rename
        
UP_pwf     <- nullp(UP_integer, id=GO_unique.genes.all, bias.data=length_vector) # make figure margins large enough for this to run...
UP_goseq   <- goseq(UP_pwf, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
UP_GO.05   <- data.frame(UP_goseq$category[UP_goseq$over_represented_pvalue<.05]) # change twice here
colnames(UP_GO.05) <- c("category")
UP_GO.05           <- merge(UP_GO.05, UP_goseq, by="category") # change here
UP_GO.05           <- UP_GO.05[order(UP_GO.05$ontology, UP_GO.05$over_represented_pvalue,-UP_GO.05$numDEInCat),] %>% 
                              dplyr::mutate(term = as.factor(term))





DOWN_integer <- as.integer(GO_unique.genes.all %in% (Frontloaded_DOWN$Airradians.TranscriptID)) # w/o day-specific ID vector
names(DOWN_integer)=GO_unique.genes.all # rename
        
DOWN_pwf     <- nullp(DOWN_integer, id=GO_unique.genes.all, bias.data=length_vector) # make figure margins large enough for this to run...
DOWN_goseq   <- goseq(DOWN_pwf, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
DOWN_GO.05   <- data.frame(DOWN_goseq$category[DOWN_goseq$over_represented_pvalue<.05]) # change twice here
colnames(DOWN_GO.05) <- c("category")
DOWN_GO.05           <- merge(DOWN_GO.05, DOWN_goseq, by="category") # change here
DOWN_GO.05           <- DOWN_GO.05[order(DOWN_GO.05$ontology, DOWN_GO.05$over_represented_pvalue,-DOWN_GO.05$numDEInCat),] %>% 
                              dplyr::mutate(term = as.factor(term))








        # # remove Biological Process GO terms with < 10 genesand  Molecular Function terms with < 2 genes in the module (with that term)
        # GO.05_filtered <- GO.05 %>% 
        #                   dplyr::mutate(numDEInCat = as.numeric(numDEInCat)) %>% 
        #                   dplyr::filter(!(numDEInCat<2 & ontology == "BP"), !(numDEInCat<2 & ontology == "MF"))
        # # write csv
        # write.csv(GO.05, file = paste("C:/Users/samjg/Documents/Github_repositories/Cvirginica_multistressor/RAnalysis/Output/WGCNA/day2_larvae/GO_analysis/GO.05",WGCNA_ColorList[i,1], "Module_unfiltered.csv", sep ='')) # save csv
        # 
        # write.csv(GO.05_filtered, file = paste("C:/Users/samjg/Documents/Github_repositories/Cvirginica_multistressor/RAnalysis/Output/WGCNA/day2_larvae/GO_analysis/GO.05",WGCNA_ColorList[i,1], "Module.csv", sep ='')) # save csv 
        # 
        # #print as we go
        # print(paste(WGCNA_ColorList[i,2], WGCNA_ColorList[i,1], "done", sep = ' '))
        
```

