---
title: "DESeq2"
author: "Samuel Gurr"
date: "2024-02-21"
output: pdf_document
---


### Set working directory
```{r setup, include=TRUE}
knitr::opts_chunk$set(echo = TRUE, 
                      warning = FALSE, 
                      message = FALSE, 
                      cache = TRUE)

# knitr::opts_knit$set(root.dir = "C:/Users/samjg/Documents/Github_repositories/Airradians_CellularMolecular_OA/RAnalysis")
knitr::opts_knit$set(root.dir = "C:/Users/samuel.gurr/Documents/Github_repositories/Airradians_CellularMolecular_OA/RAnalysis")
```

### Load libraries
```{r load_libraries, include = TRUE}
library(dplyr)
library(reshape2)
library(ggplot2)
library(kableExtra)
library(knitr)
library(tidyverse)
library(DESeq2)
library(plotrix)
library(ggpubr)
```


### Load data: choose ONE of the counts clusters (filtered raw reads or unfiltered raw reads)

### Sample metadata - Experimental treatments/groups
```{r experiment_data, include = TRUE}
# load data and read in the module membership master file 
load(file = "Output/Transcriptomics/WGCNA/challenge/Challenge-networkConstruction-stepByStep.RData")


Challenge_ModuleMembership  <- read.csv(file="Output/Transcriptomics/WGCNA/challenge/Challenge_WGCNA_ModulMembership.csv", 
                                        sep=',', header=TRUE)

# MASTER REF
# reference Cvirginica KEGG IDs (contains )
Cvirg_KEGGIDs_Ref   <- read.csv(file = "Data/Transcriptomics/metadata/Seq_Cvirg_Reference.csv",header = T)

# diamond bastx output of Airradians genome with Cvirg query
Airr_Cvirg_diamond   <- read.csv(file = "Data/Transcriptomics/metadata/seq_id_AirrCvirg_MERGED_master.csv",header = T) %>% 
                                        dplyr::rename('Cvirginica_TranscriptID' = 'blastxEval_CvirgTranscriptID') %>% 
                                        dplyr::select('Airradians_TranscriptID', 
                                                      'Cvirginica_TranscriptID')
Airr_Cvirg_Cgig_KEGG <- merge(Cvirg_KEGGIDs_Ref, Airr_Cvirg_diamond, by = 'Cvirginica_TranscriptID') %>% 
                          dplyr::rename(Airradians.TranscriptID = Airradians_TranscriptID,
                                        Protein_name = Cvirginica_Protein_name)

AirrdiansTranscript_Proteinnames <- Challenge_ModuleMembership %>% dplyr::select(Airradians.TranscriptID)
nrow(Airr_Cvirg_Cgig_KEGG) # 18841
AirrdiansTranscript_Proteinnames_2 <- (merge(AirrdiansTranscript_Proteinnames, Airr_Cvirg_Cgig_KEGG, by = "Airradians.TranscriptID", all = TRUE))


# transpose the vst challenge data used for WGCNa (jas 27 samples - truncated from sample tree (view WGCNA.Rmd)
vstExp_Challenge  <- as.data.frame(t(dds_1filt_master)) # this is our vst master file we used for WGCNA
vstExp_Challenge  <- tibble::rownames_to_column(vstExp_Challenge,"Airradians.TranscriptID") # rownames as first column
ncol(vstExp_Challenge) # 27 + AirradiansID


# write out this rlog tranformed master data (used for plotting below1) 
rlogExp_Challenge <- read.csv(file = "Output/Transcriptomics/WGCNA/challenge/Challenge_rlog_transformed.csv", sep=',', header=TRUE) %>% 
                          dplyr::select(!X)
ncol(rlogExp_Challenge) # 27 + AirradiansID



```

# For loop to assign frontloaded genes 

**About** : 
Consitutive gene expression is a ratio criteria to investigatate preparative/defensive transcription by groups 
of animals primed to a challenge. Growing evidence suggests that this response can mediate cross-tolerance to wthstand 
sudden change in the future environment. 
*Important!*: 
Although most studies point to a single direction for frontloading as 'frontloading of upregulated genes' (as described above), 
Barshis et al. 2013 and studies thereafter describe the possibility of 'frontloading of downregulated genes' as those with have 
following the transcriptome network theory, those that are consittutively expressed may permit/control a lower level of expression 
of genes (in some cases more than otherwise expressed highly!) and may further regulate the excitation of expression under an abrupt change 
In this case, tackling the frontloading of downregulated genes is of merit - although this name is very misleading 

* Frontloaded 'upregulated' genes 
- those that are in modules red and salmon show higher exp by high pco2 history > low pCO2 history
- control ratio == (HighpCO2historty[high pCO2 exposure]) : (LowpCO2History[low pCO2 exposure]), this ratio 
indicates the baseline expression level under 'home' or their current 'control' conditions
- response ratio == (HighpCO2historty[low pCO2 exposure]) / (HighpCO2historty[high pCO2 exposure]) : 
                    (LowpCO2historty[high pCO2 exposure]) / (LowpCO2historty[low pCO2 exposure]), here we assume 
that the response to change (high to low, low to high) will elicit a response, and thus within histories we want that response 
to yield >1 value such as high history's response to low employed as expoures low:high - if they respond to change then we would yield 
a value >1. Vice versa, low pCO2 history's response to high as exposures high:low. Taking the ratio of high hisotry response over low history response (to pCO2 change!), a value >1 indicates that the high history is more responsive (<1 high history is less responsive)
- frontoadeded genes are indicated by control ratio => 1 and response ratio =< 1

* Frontloaded 'downregulated' genes 
- those that are in modules green and pink show higher exp by low pco2 history > high pCO2 history
- control ratio == (HighpCO2historty[high pCO2 exposure]) : (LowpCO2History[low pCO2 exposure]), this ratio 
indicates the baseline expression level under 'home' or their current 'control' conditions
- response ratio == (HighpCO2historty[low pCO2 exposure]) / (HighpCO2historty[high pCO2 exposure]) : 
                    (LowpCO2historty[high pCO2 exposure]) / (LowpCO2historty[low pCO2 exposure]), here we assume 
that the response to change (high to low, low to high) will elicit a response, and thus within histories we want that response 
to yield >1 value such as high history's response to low employed as expoures low:high - if they respond to change then we would yield 
a value >1. Vice versa, low pCO2 history's response to high as exposures high:low. Taking the ratio of high hisotry response over low history response (to pCO2 change!), a value >1 indicates that the high history is more responsive (<1 high history is less responsive)
- **here is where they differ!** frontloaded 'downregulated' genes are indicated by control ratio =< 1 and response ratio => 1; 
such that the high pCO2 history has lower control expression but more responsive to change!


## First, filter by module ID and build read matrix for frontloading tests
(1) Upregulated genes - Modules red and salmon (N = 790)
```{r Modules red and salmon, include = TRUE}

moduleRedSalmon_Airradians.IDs       <- (Challenge_ModuleMembership %>% dplyr::filter(moduleColor %in%
                                                                                       c('red','salmon')))$Airradians.TranscriptID

moduleRedSalmon_Airradians.IDs       <- (Challenge_ModuleMembership %>% dplyr::filter(moduleColor %in%
                                                                                       c('red')))$Airradians.TranscriptID

length(moduleRedSalmon_Airradians.IDs) # 790
#vst 
vstExp_ChallengeUP.long              <- vstExp_Challenge %>% 
                                            dplyr::filter(Airradians.TranscriptID %in% moduleRedSalmon_Airradians.IDs) %>% 
                                            pivot_longer(
                                              cols = !c(Airradians.TranscriptID), 
                                              names_to = "Sample_num",
                                              values_to = "vstExp"
                                            )
nrow(vstExp_ChallengeUP.long) # 21060

vstExp_ChallengeUP.long.merge        <- merge(vstExp_ChallengeUP.long, AirrdiansTranscript_Proteinnames_2, by = 'Airradians.TranscriptID')

vstExp_ChallengeUP.long.meta         <- cbind(vstExp_ChallengeUP.long.merge, exp_metadata_1_master, by = "Sample_num")

vstExp_ChallengeUP.long.meta.means   <- vstExp_ChallengeUP.long.meta %>% 
                                            dplyr::select(c(vstExp, pCO2_exposure,pCO2_history, 
                                                            Airradians.TranscriptID, Protein_name)) %>% 
                                            dplyr::group_by(pCO2_exposure, pCO2_history, 
                                                            Airradians.TranscriptID, Protein_name) %>% 
                                            dplyr::summarise(mean.vstExp = mean(vstExp), n = n()) 

nrow(vstExp_ChallengeUP.long.meta.means) # 4680
length(unique(vstExp_ChallengeUP.long.meta.means$Airradians.TranscriptID)) # 780
```

(2) Downregulated genes - Modules green and pink (N = 1087)
```{r Modules green and pink, include = TRUE}

modulesGreenPink_Airradians.IDs      <- (Challenge_ModuleMembership %>% dplyr::filter(moduleColor %in%
                                                                                       c('green','pink')))$Airradians.TranscriptID

modulesGreenPink_Airradians.IDs      <- (Challenge_ModuleMembership %>% dplyr::filter(moduleColor %in%
                                                                                        c('green')))$Airradians.TranscriptID

length(modulesGreenPink_Airradians.IDs) # 1087
#vst 
vstExp_ChallengeDOWN.long              <- vstExp_Challenge %>% 
                                            dplyr::filter(Airradians.TranscriptID %in% modulesGreenPink_Airradians.IDs) %>% 
                                            pivot_longer(
                                              cols = !c(Airradians.TranscriptID), 
                                              names_to = "Sample_num",
                                              values_to = "vstExp"
                                            )
nrow(vstExp_ChallengeDOWN.long) # 29160

vstExp_ChallengeDOWN.long.merge        <- merge(vstExp_ChallengeDOWN.long, AirrdiansTranscript_Proteinnames_2, by = 'Airradians.TranscriptID')

vstExp_ChallengeDOWN.long.meta         <- cbind(vstExp_ChallengeDOWN.long.merge, exp_metadata_1_master, by = "Sample_num")

vstExp_ChallengeDOWN.long.meta.means   <- vstExp_ChallengeDOWN.long.meta %>% 
                                            dplyr::select(c(vstExp, pCO2_exposure,pCO2_history, 
                                                            Airradians.TranscriptID, Protein_name)) %>% 
                                            dplyr::group_by(pCO2_exposure, pCO2_history, 
                                                            Airradians.TranscriptID, Protein_name) %>% 
                                            dplyr::summarise(mean.vstExp = mean(vstExp), n = n()) 

nrow(vstExp_ChallengeDOWN.long.meta.means) # 6480
length(unique(vstExp_ChallengeDOWN.long.meta.means$Airradians.TranscriptID)) # 1080
```


## Names for the following calculations 

(1.1) Frontloading of upregulated genes **in response to severe mismatch** under high pCO2 (low history) and under low pCO2 (high history)

* Frontloaded upreuglated genes - those are the H > L globally, and we see that L upregulates in response to LxH relative to HxH
* y axis = HxL : LxL
* x axis = (H:H / H:L) : (L:H / L:L); < 1 means the LxH > HxH that the low responds to high even in the downregulated gene pool

(1.2) Frontloading of downregulated genes **in response to low mismatch** under low pCO2 (high history) and under high pCO2 (low history)
* y axis = LxH : HxH
* x axis = (L:L / L:H) : (H:L / H:H); < 1 means the HxL > LxL that the high responds to low even in the downregulated gene pool

(1.2) Frontloading of upregulated genes **in response to moderate mismatch** under moderate pCO2 (for both low and high history)

(2.1) Frontloading of downregulated genes **in response to high pCO2  mismatch** under high pCO2 (low history) and under low pCO2 (high history)

(2.2) Frontloading of donregulated genes **in response to moderate pCO2 mismatch** under moderate pCO2 (for both low and high history)

## (1.1) Frontloading of upregulated genes **in response to high pCO2 mismatch** 
```{r, 1.1 Frontloading of upregulated genes in response to high pCO2  mismatch}
loop_df    <- as.data.frame(unique(vstExp_ChallengeUP.long.meta.means$Airradians.TranscriptID))
# loop_df    <- as.data.frame(unique(rlogExp_Challenge.long.meta.means$Airradians.TranscriptID))
loop.table <- data.frame(matrix(nrow = 1, ncol = 9)) # create dataframe to save cumunalitively during for loop
colnames(loop.table)<-c('Airradians.TranscriptID',
                        'Protein_name',
                        'baseMeanNAIVE_control', 
                        'baseMeanHABITUATED_control', 
                        'baseMeanNAIVE_response', 
                        'baseMeanHABITUATED_response', 
                        'ControlRatio',
                        'ResponseRatio',
                        'Frontloaded_criteria')
df_total_1.1 <- data.frame() # start dataframe 

for (i in 1:nrow(loop_df)) {
  
  df_loop <- vstExp_ChallengeUP.long.meta.means %>% filter(Airradians.TranscriptID %in% loop_df[i,])

    if (nrow(df_loop) == 6) {
  
  loop.table$Airradians.TranscriptID     <- loop_df[i,]
  
  loop.table$Protein_name                <- df_loop$Protein_name[1]

  # needs to be a control ratio under low pCO2 to prevent redundancy in the response ratio calculation 
  # oetherwise you would divide the same numbers by one another and == 1, here we call the control ratio as the exopsure 
  # under low pCO2 for both histories, to truncate genes that have higher expression by the high history
  loop.table$baseMeanNAIVE_control       <- (df_loop %>% dplyr::filter(pCO2_exposure %in% 'low' & # low under low
                                                                         pCO2_history %in% 'low'))$mean.vstExp
  loop.table$baseMeanHABITUATED_control  <- (df_loop %>% dplyr::filter(pCO2_exposure %in% 'low' &  # high under high
                                                                         pCO2_history %in% 'severe'))$mean.vstExp
  
  loop.table$baseMeanNAIVE_response      <- (df_loop %>% dplyr::filter(pCO2_exposure %in% 'severe' & # low's response to high
                                                                         pCO2_history %in% 'low'))$mean.vstExp
  loop.table$baseMeanHABITUATED_response <- (df_loop %>% dplyr::filter(pCO2_exposure %in% 'severe' &  # high's response to low
                                                                         pCO2_history %in% 'severe'))$mean.vstExp
  
  loop.table$ControlRatio                <- loop.table$baseMeanHABITUATED_control / loop.table$baseMeanNAIVE_control

  loop.table$ResponseRatio               <- ( (loop.table$baseMeanHABITUATED_response / 
                                                loop.table$baseMeanHABITUATED_control) /
    
                                            (loop.table$baseMeanNAIVE_response / 
                                                loop.table$baseMeanNAIVE_control) )
  
  # loop.table$ResponseRatio               <- ( abs(1-(loop.table$baseMeanHABITUATED_response / 
  #                                               loop.table$baseMeanHABITUATED_control)) /
  #   
  #                                           abs(1-(loop.table$baseMeanNAIVE_response / 
  #                                               loop.table$baseMeanNAIVE_control)) )
  
  loop.table <- loop.table %>% 
                          dplyr::mutate(Frontloaded_criteria = 
                                          case_when(
                                            ControlRatio > 1 & 
                                              ResponseRatio < 1 ~ "frontloaded",
                                            
                                            ControlRatio < 1 & 
                                              ResponseRatio < 1 ~ "low expression relative to naive",
                                            
                                            ControlRatio < 1 & 
                                              ResponseRatio > 1 ~ "more responsive relative to naive"))
  } else {NA}
  df <- data.frame(loop.table) # name dataframe for this singl e row
  df_total_1.1 <- rbind(df_total_1.1, df) #bind to a cumulative list dataframe
}    
nrow(df_total_1.1) #780
Frontloading_plot_1.1 <- df_total_1.1 %>% 
                        # dplyr::filter(ControlRatio < 5) %>% 
                        # dplyr::filter(ResponseRatio  < 25) %>% 
                                      ggplot(aes(x=ResponseRatio, y=ControlRatio)) +
                                      geom_point(aes(color='grey80', na.rm=TRUE)) +  
                                         scale_shape_manual(values=c(19,19)) + 
                                         scale_color_manual(values=c("grey", "black")) +
                                      theme_classic() + 
                                      # scale_x_continuous(expand = c(0, 0), limits = c(0.9,1.1)) +
                                      # scale_y_continuous(expand = c(0, 0), limits = c(0.95,1.1)) +
                                      stat_smooth(method = "lm", 
                                                  formula = y ~ x + poly(x, 2) - 1) +
                                      geom_vline(xintercept=1, linetype="dotted") + 
                                      geom_hline(yintercept=1, linetype="dotted") + 
                                      labs(y= "Control ratio, under matched pCO2", 
                                           x = "Response ratio, under mismatched pCO2",
                                           title = "Frontloading criteria (shaded)") + 
                                      expand_limits(x = 1.1, y = 1.1) + 
                                      annotate("rect", xmin = 0.9, xmax = 1, ymin = 1, ymax = 1.1,
                                               alpha = .2) + 
                                      theme(legend.position="none", text = element_text(size=10))

# truncate df_total for only genes assigned as 'frontloaded' from module red
Frontloaded_AirradiansTranscriptIDs <- merge((df_total_1.1 %>% filter(Frontloaded_criteria == 'frontloaded')),
                                             AirrdiansTranscript_Proteinnames_2, 
                                             by = c('Protein_name','Airradians.TranscriptID'))
View(Frontloaded_AirradiansTranscriptIDs)

write.csv(Frontloaded_AirradiansTranscriptIDs, 
          file = "Output/Transcriptomics/WGCNA/challenge/Frontloading/UPmodulered_Severe.csv")


# truncate the vst Exp means (mean exp for each gene within pCO2 history * pCO2 exposure, N = 5 each)
IDtargets       <- vstExp_ChallengeUP.long.meta.means %>% 
                        dplyr::filter(Airradians.TranscriptID %in%
                        unique(Frontloaded_AirradiansTranscriptIDs$Airradians.TranscriptID))

IDtargets_stats <- IDtargets %>% 
                      dplyr::filter(pCO2_exposure %in% c('low', 'severe')) %>% 
                      dplyr::mutate(MatchMismatch = case_when(
                        (pCO2_history %in% 'low'    & pCO2_exposure %in% 'low')    ~ "match",
                        (pCO2_history %in% 'severe' & pCO2_exposure %in% 'severe') ~ "match",
                        (pCO2_history %in% 'low'    & pCO2_exposure %in% 'severe') ~ "mismatch",
                        (pCO2_history %in% 'severe' & pCO2_exposure %in% 'low')    ~ "mismatch",TRUE ~ NA
                      )) %>% 
                      dplyr::mutate(IDvar = paste0(pCO2_history,MatchMismatch))
IDtargets_stats$MatchMismatch <- as.factor(IDtargets_stats$MatchMismatch)
IDtargets_stats$IDvar         <- as.factor(IDtargets_stats$IDvar)
mod <- lm(mean.vstExp~pCO2_history*MatchMismatch,data=IDtargets_stats)

shapiro.test(resid(mod)) # p-value < 2.2e-16
mod <- scheirerRayHare(mean.vstExp~pCO2_history*MatchMismatch,data=IDtargets_stats)
#                             Df   Sum Sq      H  p.value
# pCO2_history                 1   707190 9.6761 0.001867
# MatchMismatch                1   238338 3.2611 0.070943
# pCO2_history:MatchMismatch   1   289598 3.9624 0.046527
modKW <- kruskal.test(mean.vstExp~IDvar,data=IDtargets_stats)
# Kruskal-Wallis chi-squared = 16.9, df = 3, p-value = 0.0007411
modDunn <-  dunnTest(mean.vstExp ~ IDvar,
              data=IDtargets_stats,
              method="bh")
# Comparison
# lowmatch - lowmismatch	-2.6844766	0.0072643460	0.014528692	
# lowmatch - severematch	-3.6071159	0.0003096194	0.001857717	
# lowmismatch - severematch	-0.9226392	0.3561952434	0.534292865	
# lowmatch - severemismatch	-3.4764827	0.0005080369	0.001524111	
# lowmismatch - severemismatch	-0.7920061	0.4283571185	0.514028542	
# severematch - severemismatch	0.1306331	0.8960655328	0.896065533	

plotting <- IDtargets_stats %>% group_by(pCO2_history, MatchMismatch) %>%   
                          dplyr::summarise(mean_meanvstExp = mean(mean.vstExp),
                            sd_vstExp = sd(mean.vstExp),
                            n = n(),
                            se_vstExp = sd_vstExp / sqrt(n)
                          )

nrow(Frontloaded_AirradiansTranscriptIDs) # 234,  391
Frontloading_MEAN.SE_1.1 <- plotting %>% 
        dplyr::filter(MatchMismatch %in% c('match', 'mismatch')) %>% 
             ggplot(aes(x=MatchMismatch, y = mean_meanvstExp, fill=pCO2_history)) +
                     geom_point(aes(shape=pCO2_history, fill=pCO2_history), 
                                    size = 4.5,position=position_dodge(.4)) + 
               geom_line() +
               theme_classic() +
               labs(y= "vst expression", 
                    x = "pCO2 exposure",
                    title = "Frontloaded genes (N=234)") + 
               geom_errorbar(aes(ymin=(mean_meanvstExp)-(se_vstExp), # new means and se by treatment
                                 ymax=(mean_meanvstExp)+(se_vstExp)), # new means and se by treatment
                                 width=0,position=position_dodge(.4)) # width determines the length of the end ticks

pdf("Output/Transcriptomics/WGCNA/challenge/Frontloading/UPmodulered_Severe.pdf", height=4, width =8)
ggarrange(Frontloading_plot_1.1, Frontloading_MEAN.SE_1.1, ncol=2)
dev.off()
```

## (1.2) Frontloading of downregulated genes **in response to low mismatch** under low pCO2 (high history) and under high pCO2 (low history)
```{r, 1.2 Frontloading of dwonregulated genes in response to low pCO2  mismatch}
loop_df    <- as.data.frame(unique(vstExp_ChallengeDOWN.long.meta.means$Airradians.TranscriptID))
# loop_df    <- as.data.frame(unique(rlogExp_Challenge.long.meta.means$Airradians.TranscriptID))
loop.table <- data.frame(matrix(nrow = 1, ncol = 9)) # create dataframe to save cumunalitively during for loop
colnames(loop.table)<-c('Airradians.TranscriptID',
                        'Protein_name',
                        'baseMeanNAIVE_control', 
                        'baseMeanHABITUATED_control', 
                        'baseMeanNAIVE_response', 
                        'baseMeanHABITUATED_response', 
                        'ControlRatio',
                        'ResponseRatio',
                        'Frontloaded_criteria')
df_total_1.2 <- data.frame() # start dataframe 

for (i in 1:nrow(loop_df)) {
  
  df_loop <- vstExp_ChallengeDOWN.long.meta.means %>% filter(Airradians.TranscriptID %in% loop_df[i,])

    if (nrow(df_loop) == 6) {
  
  loop.table$Airradians.TranscriptID     <- loop_df[i,]
  
  loop.table$Protein_name                <- df_loop$Protein_name[1]

  # I ran this frontloaded suppressed analysis based on response the response to high pCO2 animals to low pcO2 
  # however the results were non significant in the anova model, and were generally unreliable - look at the WGCNA results, here 
  # we see that the expression level of high history to low exposure decreased, not increased, and this decline was near thesame 
  # for the decline exhibited by the low history in response to high (in a mismatch framework) - in looking at these data I noticed that 
  # the high history actually increass expressed upon moderate exposure, 
  # Below is frontloading suppression of the high pCO2 history - more responsive upon pCO2 change 
  # This is a mismatch model CHANGE = a decline for high history and an INCREASE for the low history! 
  
  loop.table$baseMeanNAIVE_control       <- (df_loop %>% dplyr::filter(pCO2_exposure %in% 'severe' & # severe under severe - they are naive to low
                                                                         pCO2_history %in% 'severe'))$mean.vstExp
  loop.table$baseMeanHABITUATED_control  <- (df_loop %>% dplyr::filter(pCO2_exposure %in% 'low' &  # low under severe - the habiturated control - low's response to severe, must be > severexsevere
                                                                         pCO2_history %in% 'low'))$mean.vstExp
  
  loop.table$baseMeanNAIVE_response      <- (df_loop %>% dplyr::filter(pCO2_exposure %in% 'moderate' & #  severe's response to low 
                                                                         pCO2_history %in% 'severe'))$mean.vstExp
  loop.table$baseMeanHABITUATED_response <- (df_loop %>% dplyr::filter(pCO2_exposure %in% 'moderate' &  # lows response to low
                                                                         pCO2_history %in% 'low'))$mean.vstExp
  
  loop.table$ControlRatio                <- loop.table$baseMeanHABITUATED_control / loop.table$baseMeanNAIVE_control
 
  loop.table$ResponseRatio               <- ( (loop.table$baseMeanHABITUATED_response / 
                                                loop.table$baseMeanHABITUATED_control) /
    
                                            (loop.table$baseMeanNAIVE_response / # THINK! we wonder if the response REDUCES expression
                                                loop.table$baseMeanNAIVE_control) ) # resonse as denomicator increases the number
  
  # loop.table$ResponseRatio               <- ( abs(1-(loop.table$baseMeanHABITUATED_response / 
  #                                               loop.table$baseMeanHABITUATED_control)) /
  #   
  #                                           abs(1-(loop.table$baseMeanNAIVE_response / 
  #                                               loop.table$baseMeanNAIVE_control)) )
  
  loop.table <- loop.table %>% 
                          dplyr::mutate(Frontloaded_criteria = 
                                          case_when(
                                            ControlRatio > 1 & 
                                              ResponseRatio < 1 ~ "frontloaded",
                                            
                                            ControlRatio < 1 & 
                                              ResponseRatio < 1 ~ "low expression relative to naive",
                                            
                                            ControlRatio < 1 & 
                                              ResponseRatio > 1 ~ "more responsive relative to naive"))
  } else {NA}
  df <- data.frame(loop.table) # name dataframe for this singl e row
  df_total_1.2 <- rbind(df_total_1.2, df) #bind to a cumulative list dataframe
}    
nrow(df_total_1.2) #690
Frontloading_plot_1.2 <- df_total_1.2 %>% 
                        dplyr::filter(!ResponseRatio  > 1.06) %>% 
                                      ggplot(aes(x=ResponseRatio, y=ControlRatio)) +
                                      geom_point(aes(color='grey80', na.rm=TRUE)) +  
                                         scale_shape_manual(values=c(19,19)) + 
                                         scale_color_manual(values=c("grey", "black")) +
                                      theme_classic() + 
                                      # scale_x_continuous(expand = c(0, 0), limits = c(0.9,1.1)) +
                                      # scale_y_continuous(expand = c(0, 0), limits = c(0.95,1.1)) +
                                      stat_smooth(method = "lm", 
                                                  formula = y ~ x + poly(x, 2) - 1) +
                                      geom_vline(xintercept=1, linetype="dotted") + 
                                      geom_hline(yintercept=1, linetype="dotted") + 
                                      labs(y= "Control ratio, under matched pCO2", 
                                           x = "Response ratio, under mismatched pCO2",
                                           title = "Frontloading criteria (shaded)") + 
                                      expand_limits(x = 1.1, y = 1.1) + 
                                      annotate("rect", xmin = 0.9, xmax = 1, ymin = 1, ymax = 1.1,
                                               alpha = .2) + 
                                      theme(legend.position="none", text = element_text(size=10))

# truncate df_total for only genes assigned as 'frontloaded' from module red
Frontloaded_AirradiansTranscriptIDs <- merge((df_total_1.2 %>% filter(Frontloaded_criteria == 'frontloaded')),
                                             AirrdiansTranscript_Proteinnames_2, 
                                             by = c('Protein_name','Airradians.TranscriptID'))
nrow(Frontloaded_AirradiansTranscriptIDs) # 324
# View(Frontloaded_AirradiansTranscriptIDs)
write.csv(Frontloaded_AirradiansTranscriptIDs, 
          file = "Output/Transcriptomics/WGCNA/challenge/Frontloading/DOWNmoduleGreen_Moderate.csv")


# truncate the vst Exp means (mean exp for each gene within pCO2 history * pCO2 exposure, N = 5 each)
IDtargets       <- vstExp_ChallengeDOWN.long.meta.means %>% 
                        dplyr::filter(Airradians.TranscriptID %in%
                        unique(Frontloaded_AirradiansTranscriptIDs$Airradians.TranscriptID))
nrow(IDtargets) / 6 # 312 - about the exact  size of the Frontloaded_AirradiansTranscriptIDs

IDtargets_stats <- IDtargets %>% 
                      dplyr::filter(pCO2_exposure %in% c('low', 'moderate', 'severe')) %>% 
                      dplyr::filter(!pCO2_history %in% 'moderate') %>% 
                      dplyr::mutate(MatchMismatch = case_when(
                        (pCO2_history %in% 'severe'  & pCO2_exposure %in% 'severe')    ~ "match",
                        (pCO2_history %in% 'low' & pCO2_exposure %in% 'low') ~ "match",
                        (pCO2_history %in% 'severe'  & pCO2_exposure %in% 'moderate') ~ "mismatch",
                        (pCO2_history %in% 'low' & pCO2_exposure %in% 'moderate')    ~ "mismatch",TRUE ~ NA
                      )) %>%  na.omit() %>% 
                      dplyr::mutate(IDvar = paste0(pCO2_history,MatchMismatch))
IDtargets_stats$IDvar <- as.factor(IDtargets_stats$IDvar)
unique(IDtargets_stats$IDvar)
# mod <- lm(mean.vstExp~pCO2_history*pCO2_exposure,data=IDtargets_stats)
mod <- lm(mean.vstExp~IDvar,data=IDtargets_stats)
shapiro.test(resid(mod)) # p-value < 2.2e-16
library(rcompanion)
library(FSA)
scheirerRayHare(mean.vstExp~pCO2_history*MatchMismatch,data=IDtargets_stats)
modKW   <- kruskal.test(mean.vstExp~IDvar,data=IDtargets_stats)
# Kruskal-Wallis chi-squared = 15.477, df = 3, p-value = 0.001451
modDunn <-  dunnTest(mean.vstExp ~ IDvar,
              data=IDtargets_stats,
              method="bh")
# Comparison
# lowmatch - lowmismatch	0.2238306	0.8228891338	0.822889134	
# lowmatch - severematch	3.5053043	0.0004560857	0.002736514	
# lowmismatch - severematch	3.2814737	0.0010326616	0.003097985	
# lowmatch - severemismatch	1.4644960	0.1430584631	0.214587695	
# lowmismatch - severemismatch	1.2406655	0.2147293575	0.257675229	
# severematch - severemismatch	-2.0408082	0.0412698919	0.082539784	


plotting <- IDtargets_stats %>% group_by(pCO2_history, MatchMismatch) %>%   
                          dplyr::summarise(mean_meanvstExp = mean(mean.vstExp),
                            sd_vstExp = sd(mean.vstExp),
                            n = n(),
                            se_vstExp = sd_vstExp / sqrt(n)
                          )

nrow(Frontloaded_AirradiansTranscriptIDs) # 324
Frontloading_MEAN.SE_1.2 <- plotting %>% 
        dplyr::filter(MatchMismatch %in% c('match', 'mismatch')) %>% 
             ggplot(aes(x=MatchMismatch, y = mean_meanvstExp, fill=pCO2_history)) +
                     geom_point(aes(shape=pCO2_history, fill=pCO2_history), 
                                    size = 4.5,position=position_dodge(.4)) + 
               geom_line() +
               theme_classic() +
               labs(y= "vst expression", 
                    x = "pCO2 exposure",
                    title = "Frontloaded genes (N=324)") + 
               geom_errorbar(aes(ymin=(mean_meanvstExp)-(se_vstExp), # new means and se by treatment
                                 ymax=(mean_meanvstExp)+(se_vstExp)), # new means and se by treatment
                                 width=0,position=position_dodge(.4)) # width determines the length of the end ticks


pdf("Output/Transcriptomics/WGCNA/challenge/Frontloading/DOWNGreen_Moderate.pdf", height=4, width =8)
ggarrange(Frontloading_plot_1.2, Frontloading_MEAN.SE_1.2, ncol=2)
dev.off()
```
## (1.3) Frontloading of upregulated genes **in response to moderate pCO2 mismatch** 
```{r, 1.2 Frontloading of upregulated genes in response to moderate pCO2 mismatch}
loop_df    <- as.data.frame(unique(vstExp_Challenge.long.meta.means$Airradians.TranscriptID))
# loop_df    <- as.data.frame(unique(rlogExp_Challenge.long.meta.means$Airradians.TranscriptID))
loop.table <- data.frame(matrix(nrow = 1, ncol = 9)) # create dataframe to save cumunalitively during for loop
colnames(loop.table)<-c('Airradians.TranscriptID',
                        'Protein_name',
                        'baseMeanNAIVE_control', 
                        'baseMeanHABITUATED_control', 
                        'baseMeanNAIVE_response', 
                        'baseMeanHABITUATED_response', 
                        'ControlRatio',
                        'ResponseRatio',
                        'Frontloaded_criteria')
df_total_1.2 <- data.frame() # start dataframe 


for (i in 1:nrow(loop_df)) {
  
  df_loop <- vstExp_ChallengeUP.long.meta.means %>% filter(Airradians.TranscriptID %in% loop_df[i,])

    if (nrow(df_loop) == 6) {
  
  loop.table$Airradians.TranscriptID     <- loop_df[i,]
  
  loop.table$Protein_name                <- df_loop$Protein_name[1]

  loop.table$baseMeanNAIVE_control       <- (df_loop %>% dplyr::filter(pCO2_exposure %in% 'low' & 
                                                                         pCO2_history %in% 'low'))$mean.vstExp
  loop.table$baseMeanHABITUATED_control  <- (df_loop %>% dplyr::filter(pCO2_exposure %in% 'low' & 
                                                                         pCO2_history %in% 'severe'))$mean.vstExp
  
  loop.table$baseMeanNAIVE_response      <- (df_loop %>% dplyr::filter(pCO2_exposure %in% 'moderate' & 
                                                                         pCO2_history %in% 'low'))$mean.vstExp
  loop.table$baseMeanHABITUATED_response <- (df_loop %>% dplyr::filter(pCO2_exposure %in% 'moderate' & 
                                                                         pCO2_history %in% 'severe'))$mean.vstExp
  
  loop.table$ControlRatio                <- loop.table$baseMeanHABITUATED_control / loop.table$baseMeanNAIVE_control
  
  loop.table$ResponseRatio               <- ( (loop.table$baseMeanHABITUATED_response / 
                                                loop.table$baseMeanHABITUATED_control) /
    
                                              (loop.table$baseMeanNAIVE_response /
                                                loop.table$baseMeanNAIVE_control) )
  
  loop.table <- loop.table %>% 
                          dplyr::mutate(Frontloaded_criteria = 
                                          case_when(
                                            ControlRatio > 1 & 
                                              ResponseRatio < 1 ~ "frontloaded",
                                            
                                            ControlRatio < 1 & 
                                              ResponseRatio < 1 ~ "low expression relative to naive",
                                            
                                            ControlRatio < 1 & 
                                              ResponseRatio > 1 ~ "more responsive relative to naive"))
  } else {NA}
  df <- data.frame(loop.table) # name dataframe for this singl e row
  df_total_1.2 <- rbind(df_total_1.2, df) #bind to a cumulative list dataframe
}    
#Plot the criteria 
Frontloading_plot_1.2 <- df_total_1.2 %>% 
                        dplyr::mutate(FrontMod_color = 
                                        ifelse(ResponseRatio < 1.0 & 
                                                 ControlRatio > 1.0, "True", "False")) %>%
                        # dplyr::filter(ControlRatio < 5) %>% 
                        # dplyr::filter(ResponseRatioHigh <1.2) %>% 
                                      ggplot(aes(x=ResponseRatio, y=ControlRatio)) +
                                      geom_point(aes(color='grey80', na.rm=TRUE)) +  
                                         scale_shape_manual(values=c(19,19)) + 
                                         scale_color_manual(values=c("grey", "black")) +
                                      theme_classic() + 
                                      # scale_x_continuous(expand = c(0, 0), limits = c(0.9,1.1)) +
                                      # scale_y_continuous(expand = c(0, 0), limits = c(0.95,1.1)) +
                                      stat_smooth(method = "lm", 
                                                  formula = y ~ x + poly(x, 2) - 1) +
                                      geom_vline(xintercept=1, linetype="dotted") + 
                                      geom_hline(yintercept=1, linetype="dotted") + 
                                      labs(y= "Control ratio, under matched pCO2", 
                                           x = "Response ratio, under mismatched pCO2",
                                           title = "Frontloading criteria (shaded)") + 
                                      expand_limits(x = 1.1, y = 1.1) + 
                                      annotate("rect", xmin = 0.9, xmax = 1, ymin = 1, ymax = 1.1,
                                               alpha = .2) + 
                                      theme(legend.position="none", text = element_text(size=10))


# call the unique frontloaded genes
WGCNA_frontloaded_1.2 <- unique(merge((df_total_1.2 %>% filter(Frontloaded_criteria == 'frontloaded')), 
                            AirrdiansTranscript_Proteinnames_2, by = 'Airradians.TranscriptID'))


# truncate df_total for only genes assigned as 'frontloaded' from modules red and salmon
Frontloaded_AirradiansTranscriptIDs <- merge((df_total_1.2 %>% filter(Frontloaded_criteria == 'frontloaded')),
                                             AirrdiansTranscript_Proteinnames_2, 
                                             by = c('Protein_name','Airradians.TranscriptID'))

write.csv(Frontloaded_AirradiansTranscriptIDs, 
          file = "Output/Transcriptomics/WGCNA/challenge/Frontloading/UPmoduleredsalmon_Moderate.csv")



# truncate the vst Exp means (mean exp for each gene within pCO2 history * pCO2 exposure, N = 5 each)
IDtargets                           <- vstExp_ChallengeUP.long.meta.means %>% 
                                          dplyr::filter(Airradians.TranscriptID %in%
                                          unique(Frontloaded_AirradiansTranscriptIDs$Airradians.TranscriptID))

IDtargets_stats <- IDtargets %>% 
                      dplyr::mutate(MatchMismatch = case_when(
                        (pCO2_history %in% 'low'    & pCO2_exposure %in% 'low')    ~ "match",
                        (pCO2_history %in% 'severe' & pCO2_exposure %in% 'severe') ~ "match",
                        (pCO2_history %in% 'low'    & pCO2_exposure %in% 'moderate') ~ "mismatch",
                        (pCO2_history %in% 'severe' & pCO2_exposure %in% 'moderate')    ~ "mismatch",TRUE ~ NA
                      )) %>% 
                      dplyr::filter(MatchMismatch %in% c('match', 'mismatch'))

mod <- lm(mean.vstExp~pCO2_history*MatchMismatch,data=IDtargets_stats)
shapiro.test(resid(mod)) # p-value < 2.2e-16
mod <- scheirerRayHare(mean.vstExp~pCO2_history*MatchMismatch,data=IDtargets_stats)
mod

plotting <- IDtargets_stats %>% group_by(pCO2_history, MatchMismatch) %>%   
                          dplyr::summarise(mean_meanvstExp = mean(mean.vstExp),
                            sd_vstExp = sd(mean.vstExp),
                            n = n(),
                            se_vstExp = sd_vstExp / sqrt(n)
                          )

nrow(Frontloaded_AirradiansTranscriptIDs) # 300
Frontloading_MEAN.SE_1.2 <-  plotting %>% 
        dplyr::filter(MatchMismatch %in% c('match', 'mismatch')) %>% 
             ggplot(aes(x=MatchMismatch, y = mean_meanvstExp, fill=pCO2_history)) +
                     geom_point(aes(shape=pCO2_history, fill=pCO2_history), 
                                    size = 4.5,position=position_dodge(.4)) + 
               geom_line() +
               theme_classic() +
               labs(y= "vst expression", 
                    x = "pCO2 exposure",
                    title = "Frontloaded genes (N=391)") + 
               geom_errorbar(aes(ymin=(mean_meanvstExp)-(se_vstExp), # new means and se by treatment
                                 ymax=(mean_meanvstExp)+(se_vstExp)), # new means and se by treatment
                                 width=0,position=position_dodge(.4)) # width determines the length of the end ticks

pdf("Output/Transcriptomics/WGCNA/challenge/Frontloading/UPmoduleredsalmon_Moderate.pdf", height=4, width =8)
ggarrange(Frontloading_plot_1.2, Frontloading_MEAN.SE_1.2, ncol=2)
dev.off()
```

## (2.2) Frontloading of downrgulated genes **in response to moderate pCO2 mismatch** 
```{r, 2.2 Frontloading of upregulated genes in response to moderate pCO2 mismatch}
loop_df    <- as.data.frame(unique(vstExp_ChallengeDOWN.long.meta.means$Airradians.TranscriptID))
# loop_df    <- as.data.frame(unique(rlogExp_Challenge.long.meta.means$Airradians.TranscriptID))
loop.table <- data.frame(matrix(nrow = 1, ncol = 9)) # create dataframe to save cumunalitively during for loop
colnames(loop.table)<-c('Airradians.TranscriptID',
                        'Protein_name',
                        'baseMeanNAIVE_control', 
                        'baseMeanHABITUATED_control', 
                        'baseMeanNAIVE_response', 
                        'baseMeanHABITUATED_response', 
                        'ControlRatio',
                        'ResponseRatio',
                        'Frontloaded_criteria')
df_total_2.2 <- data.frame() # start dataframe 


for (i in 1:nrow(loop_df)) {
  
  df_loop <- vstExp_ChallengeDOWN.long.meta.means %>% filter(Airradians.TranscriptID %in% loop_df[i,])

    if (nrow(df_loop) == 6) {
  
  loop.table$Airradians.TranscriptID     <- loop_df[i,]
  
  loop.table$Protein_name                <- df_loop$Protein_name[1]

  # contrl - their matched condition
  loop.table$baseMeanNAIVE_control       <- (df_loop %>% dplyr::filter(pCO2_exposure %in% 'low' & 
                                                                         pCO2_history %in% 'low'))$mean.vstExp
  loop.table$baseMeanHABITUATED_control  <- (df_loop %>% dplyr::filter(pCO2_exposure %in% 'severe' & 
                                                                         pCO2_history %in% 'severe'))$mean.vstExp
  # response - their mismatched condition
  loop.table$baseMeanNAIVE_response      <- (df_loop %>% dplyr::filter(pCO2_exposure %in% 'moderate' & # how do low respond to high
                                                                         pCO2_history %in% 'low'))$mean.vstExp
  loop.table$baseMeanHABITUATED_response <- (df_loop %>% dplyr::filter(pCO2_exposure %in% 'moderate' & # how to high responds to low
                                                                         pCO2_history %in% 'severe'))$mean.vstExp
  
  # reverse it for down reg- naive over the habituated 
  # > 1 = higher expression by naive animals in response their control/ matched condition, lower baseline expression by habituated
  # the vast majoirty are going to be here given this is the main criteria for choosing green and pink modules
  loop.table$ControlRatio                <- loop.table$baseMeanNAIVE_control / loop.table$baseMeanHABITUATED_control
  
 # reverse it for downreg -  Naive over habituated
  # < 1 means that the habituated animals have a stronger response to mismatch 
  # loop.table$ResponseRatio           <- ( abs(1-(loop.table$baseMeanNAIVE_response / 
  #                                               loop.table$baseMeanNAIVE_control)) /
  #   
  #                                           abs(1-(loop.table$baseMeanHABITUATED_response / 
  #                                               loop.table$baseMeanHABITUATED_control)) )
  
  loop.table$ResponseRatio           <- ( (loop.table$baseMeanNAIVE_response / 
                                                loop.table$baseMeanNAIVE_control) /
    
                                            (loop.table$baseMeanHABITUATED_response / 
                                                loop.table$baseMeanHABITUATED_control) )  
  loop.table <- loop.table %>% 
                          dplyr::mutate(Frontloaded_criteria = 
                                          case_when(
                                            ControlRatio > 1 & 
                                              ResponseRatio > 1 ~ "frontloaded",
                                            
                                            ControlRatio < 1 & 
                                              ResponseRatio < 1 ~ "high expression relative to naive"))
  } else {NA}
  df <- data.frame(loop.table) # name dataframe for this singl e row
  df_total_2.2 <- rbind(df_total_2.2, df) #bind to a cumulative list dataframe
}    
#Plot the criteria 
Frontloading_plot_2.2 <- df_total_2.2 %>% 
                        dplyr::mutate(FrontMod_color = 
                                        ifelse(ResponseRatio > 1.0 & 
                                                 ControlRatio < 1.0, "True", "False")) %>%
                        # dplyr::filter(ControlRatio < 5) %>% 
                        # dplyr::filter(ResponseRatioHigh <1.2) %>% 
                                      ggplot(aes(x=ResponseRatio, y=ControlRatio)) +
                                      geom_point(aes(color='grey80', na.rm=TRUE)) +  
                                         scale_shape_manual(values=c(19,19)) + 
                                         scale_color_manual(values=c("grey", "black")) +
                                      theme_classic() + 
                                      scale_x_continuous(expand = c(0, 0), limits = c(0.9,1.1)) +
                                      scale_y_continuous(expand = c(0, 0), limits = c(0.95,1.1)) +
                                      stat_smooth(method = "lm", 
                                                  formula = y ~ x + poly(x, 2) - 1) +
                                      geom_vline(xintercept=1, linetype="dotted") + 
                                      geom_hline(yintercept=1, linetype="dotted") + 
                                      labs(y= "Control ratio, under matched pCO2", 
                                           x = "Response ratio, under mismatched pCO2",
                                           title = "Frontloading criteria (shaded)") + 
                                      expand_limits(x = 1.1, y = 1.1) + 
                                      annotate("rect", xmin = .9, xmax = 1, ymin = 1, ymax = 1.1,
                                               alpha = .2) + 
                                      theme(legend.position="none", text = element_text(size=10))


# call the unique frontloaded genes
WGCNA_frontloaded_2.2 <- unique(merge((df_total_2.2 %>% filter(Frontloaded_criteria == 'frontloaded')), 
                            AirrdiansTranscript_Proteinnames_2, by = 'Airradians.TranscriptID'))


# truncate df_total for only genes assigned as 'frontloaded' from modules red and salmon
Frontloaded_AirradiansTranscriptIDs <- merge((df_total_2.2 %>% filter(Frontloaded_criteria == 'frontloaded')),
                                             AirrdiansTranscript_Proteinnames_2, 
                                             by = c('Protein_name','Airradians.TranscriptID'))

write.csv(Frontloaded_AirradiansTranscriptIDs, 
          file = "Output/Transcriptomics/WGCNA/challenge/Frontloading/DOWNmodulegreenpink_Moderate.csv")



# truncate the vst Exp means (mean exp for each gene within pCO2 history * pCO2 exposure, N = 5 each)
IDtargets                           <- vstExp_ChallengeDOWN.long.meta.means %>% 
                                          dplyr::filter(Airradians.TranscriptID %in%
                                          unique(Frontloaded_AirradiansTranscriptIDs$Airradians.TranscriptID))
library(Rmisc)
library(rcompanion)
?scheirerRayHare
IDtargets_stats <- IDtargets %>% 
                        dplyr::mutate(MatchMismatch = case_when(
                        (pCO2_history %in% 'low'    & pCO2_exposure %in% 'low')    ~ "match",
                        (pCO2_history %in% 'severe' & pCO2_exposure %in% 'severe') ~ "match",
                        (pCO2_history %in% 'low'    & pCO2_exposure %in% 'moderate') ~ "mismatch",
                        (pCO2_history %in% 'severe' & pCO2_exposure %in% 'moderate') ~ "mismatch",TRUE ~ NA
                      )) %>% 
                      dplyr::filter(MatchMismatch %in% c('match', 'mismatch'))
AOVmod <- lm(mean.vstExp~pCO2_history*MatchMismatch,data=IDtargets_stats)
shapiro.test(resid(AOVmod)) # p-value < 2.2e-16
library(rcompanion)
SRHmod <- scheirerRayHare(mean.vstExp~pCO2_history*MatchMismatch,data=IDtargets_stats)
#                              Df    Sum Sq       H  p.value
# pCO2_history                  1   4351662 18.2297 0.000020
# MatchMismatch                 1    579975  2.4296 0.119064
# pCO2_history:MatchMismatch    1   1026924  4.3019 0.038069
# Residuals                  1688 397705121
### Dunn test
library(FSA)


plotting <- IDtargets_stats %>% group_by(pCO2_history, MatchMismatch) %>%   
                          dplyr::summarise(mean_meanvstExp = mean(mean.vstExp),
                            sd_vstExp = sd(mean.vstExp),
                            n = n(),
                            se_vstExp = sd_vstExp / sqrt(n)
                          )

nrow(Frontloaded_AirradiansTranscriptIDs) # 438
Frontloading_MEAN.SE_2.2 <- plotting %>% 
          # dplyr::filter(pCO2_exposure %in% c('low', 'moderate')) %>% 
           dplyr::filter(MatchMismatch %in% c('match', 'mismatch')) %>% 
             ggplot(aes(x=MatchMismatch, y = mean_meanvstExp, fill=pCO2_history)) +
                     geom_point(aes(shape=pCO2_history, fill=pCO2_history), 
                                    size = 4.5,position=position_dodge(.4)) + 
               geom_line() +
               theme_classic() +
               labs(y= "vst expression", 
                    x = "pCO2 exposure",
                    title = "Frontloaded genes (N = 438)") + 
               geom_errorbar(aes(ymin=(mean_meanvstExp)-(se_vstExp), # new means and se by treatment
                                 ymax=(mean_meanvstExp)+(se_vstExp)), # new means and se by treatment
                                 width=0,position=position_dodge(.4)) # width determines the length of the end ticks

pdf("Output/Transcriptomics/WGCNA/challenge/Frontloading/DOWNmodulegreenpink_Moderate.pdf", height=4, width =8)
ggarrange(Frontloading_plot_2.2, Frontloading_MEAN.SE_2.2, ncol=2)
dev.off()
```

